<a href="https://colab.research.google.com/github/Adifasyahira/Machine-Learning/blob/main/UAS%20Machine%20Learning/00_pytorch_fundamentals.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import torch
torch.__version__

'2.1.0+cu121'

torch.__version__ adalah potongan kode dalam Python yang digunakan untuk mendapatkan informasi versi dari perpustakaan PyTorch yang terpasang di lingkungan Anda. Ketika dieksekusi, potongan kode ini akan mengembalikan string yang mewakili nomor versi dari PyTorch yang Anda gunakan. Misalnya, jika Anda menjalankan kode ini dan mendapatkan keluaran seperti 1.9.0, itu berarti Anda menggunakan PyTorch versi 1.9.0.

Setiap versi PyTorch mungkin memiliki fitur, perbaikan, dan peningkatan yang berbeda. Penting untuk mengetahui versi yang Anda gunakan karena fungsionalitas atau sintaksis dapat berubah antar versi, dan beberapa kode mungkin berperilaku berbeda atau memerlukan penyesuaian berdasarkan versi PyTorch yang digunakan.

In [ ]:
# Scalar
scalar = torch.tensor(7)
scalar

tensor(7)

membuat variabel scalar yang merupakan sebuah tensor dengan hanya satu elemen. Tensor ini merepresentasikan sebuah skalar, yang merupakan bilangan tunggal. Dalam hal ini, nilai skalar yang digunakan adalah 7.

PyTorch memiliki konsep tensor yang dapat memiliki berbagai dimensi, mulai dari skalar (tanpa dimensi atau 0 dimensi) hingga tensor dengan banyak dimensi. Skalar, dalam hal ini, adalah tensor yang paling sederhana dengan hanya satu nilai.

Dengan menggunakan torch.tensor(7), Anda telah membuat tensor skalar yang dapat digunakan untuk operasi matematika, perhitungan, atau tugas-tugas lain yang memerlukan representasi dari sebuah nilai tunggal dalam lingkungan PyTorch.

In [ ]:
scalar.ndim

0

Pada PyTorch, scalar yang Anda buat adalah tensor dengan hanya satu elemen, yang pada dasarnya adalah sebuah skalar. Namun, karena ini adalah skalar, ia tidak memiliki dimensi dalam arti konvensional.

Konsep dimensi dalam tensor biasanya mengacu pada jumlah sumbu atau dimensi yang ada dalam tensor. Misalnya, tensor 1D memiliki satu sumbu, tensor 2D memiliki dua sumbu, dan seterusnya. Tetapi pada kasus skalar, karena hanya ada satu nilai tunggal tanpa arah atau dimensi tambahan, scalar tidak memiliki dimensi yang dapat dihitung menggunakan ndim.

Pemanggilan scalar.ndim pada skalar tidak akan memberikan nilai dimensi karena skalar tidak memiliki dimensi yang bisa diukur dalam konteks tensor. Dalam PyTorch, untuk skalar, nilai ndim akan menghasilkan nilai 0 karena itu bukan tensor dengan dimensi apapun.

In [ ]:
# Get the Python number within a tensor (only works with one-element tensors)
scalar.item()

7

Potongan kode scalar.item() digunakan untuk mendapatkan nilai skalar yang terkandung dalam tensor yang hanya memiliki satu elemen. Jika Anda memiliki tensor yang mewakili sebuah skalar (seperti yang Anda definisikan sebelumnya dengan torch.tensor(7)), Anda dapat menggunakan metode item() untuk mengakses nilai skalar Python yang sesungguhnya.

Contohnya, ketika Anda memiliki tensor scalar yang merupakan skalar dengan nilai 7, menggunakan scalar.item() akan mengembalikan nilai Python 7 (dalam bentuk objek Python biasa, bukan tensor PyTorch).

Hal ini berguna ketika Anda perlu menggunakan nilai dalam operasi Python biasa, karena item() mengonversi nilai dari tensor PyTorch menjadi tipe data Python yang sesuai, misalnya, dari tensor skalar menjadi integer dalam hal ini. Penting untuk diingat bahwa metode item() hanya berfungsi pada tensor yang memiliki tepat satu elemen.

In [ ]:
# Vector
vector = torch.tensor([7, 7])
vector

tensor([7, 7])

membuat variabel vector yang merupakan sebuah tensor. Dalam hal ini, tensor tersebut adalah vektor. Vektor adalah struktur data dalam matematika yang berisi kumpulan nilai, yang disusun dalam satu dimensi.

Dalam kode ini, torch.tensor([7, 7]) menciptakan sebuah vektor dengan dua elemen, yaitu 7 dan 7. Vektor tersebut merupakan vektor 1D karena terdiri dari satu baris nilai. Setiap elemen vektor dapat diakses dengan indeksnya, di mana indeks dimulai dari 0. Jadi, dalam vektor ini, vector[0] akan mengembalikan nilai 7 (elemen pertama) dan vector[1] akan mengembalikan nilai 7 (elemen kedua).

Vektor merupakan salah satu bentuk yang penting dalam komputasi numerik dan machine learning, sering kali digunakan untuk merepresentasikan fitur, prediksi, atau output dalam berbagai algoritma machine learning.

In [ ]:
# Check the number of dimensions of vector
vector.ndim

1

Maaf atas kebingungan sebelumnya. Dalam PyTorch, tidak ada properti ndim untuk langsung mendapatkan jumlah dimensi dari tensor. Sebagai gantinya, Anda dapat menggunakan fungsi dim() atau atribut dim untuk mendapatkan jumlah dimensi dari tensor.

Dalam kasus vektor yang Anda miliki, yang dibuat dengan torch.tensor([7, 7]), Anda dapat menggunakan atribut dim atau fungsi dim() untuk mengetahui jumlah dimensinya.

python
Copy code
vector.dim()  # atau len(vector.shape)
Untuk tensor vektor, ini akan mengembalikan nilai 1 karena vektor dalam PyTorch merupakan tensor dengan satu dimensi. Jumlah dimensi ini dapat penting dalam operasi-operasi tensor yang lebih kompleks atau dalam menentukan kompatibilitas antar tensor dalam berbagai operasi matematika atau pengolahan data.

In [ ]:
# Check shape of vector
vector.shape

torch.Size([2])

Dalam PyTorch, atribut shape digunakan untuk mengetahui struktur atau bentuk (shape) dari tensor. Untuk vektor yang telah Anda definisikan sebelumnya, yaitu vector = torch.tensor([7, 7]), vector.shape akan memberikan informasi tentang jumlah elemen dalam setiap dimensi tensor.

Untuk vektor tersebut, vector.shape akan mengembalikan sebuah tupel yang menunjukkan jumlah elemen dalam setiap dimensi. Dalam kasus vektor, ini akan menghasilkan (2,), yang berarti bahwa tensor ini memiliki 2 elemen dalam dimensi pertamanya (1 dimensi), sesuai dengan vektor yang hanya memiliki satu baris nilai.

Dalam kasus ini, vektor memiliki 2 elemen, dan dengan hanya memiliki satu dimensi, bentuknya direpresentasikan sebagai tupel dengan satu elemen, yaitu (2,).

In [ ]:
# Matrix
MATRIX = torch.tensor([[7, 8],
                       [9, 10]])
MATRIX

tensor([[ 7,  8],
        [ 9, 10]])

Potongan kode ini membuat variabel MATRIX yang merupakan sebuah tensor. Lebih tepatnya, ini adalah sebuah matriks. Matriks adalah struktur data dalam matematika yang memiliki baris dan kolom, yang biasanya digunakan untuk merepresentasikan data dua dimensi.

Dalam kode ini, torch.tensor([[7, 8], [9, 10]]) menciptakan sebuah matriks dengan dua baris dan dua kolom. Elemen-elemennya adalah:

lua
Copy code
[[7, 8],
 [9, 10]]
Ini adalah matriks 2x2, yang berarti memiliki dua baris dan dua kolom. Dalam matriks ini, MATRIX[0, 0] akan mengembalikan nilai 7 (baris pertama, kolom pertama), MATRIX[0, 1] akan mengembalikan nilai 8 (baris pertama, kolom kedua), MATRIX[1, 0] akan mengembalikan nilai 9 (baris kedua, kolom pertama), dan MATRIX[1, 1] akan mengembalikan nilai 10 (baris kedua, kolom kedua).

Matriks digunakan luas dalam matematika, fisika, teknik, dan juga dalam berbagai bidang ilmu seperti dalam komputasi numerik dan machine learning untuk merepresentasikan data yang memiliki hubungan dua dimensi.

In [ ]:
# Check number of dimensions
MATRIX.ndim

2

Untuk mengetahui jumlah dimensi dari tensor MATRIX yang telah Anda buat, Anda bisa menggunakan metode .dim() atau juga .ndimension() pada tensor PyTorch.

python
Copy code
MATRIX.dim()  # atau MATRIX.ndimension()
Tensor MATRIX yang merupakan matriks 2x2 dalam kasus ini akan memiliki 2 dimensi. Ini karena matriks memiliki struktur dua dimensi: satu dimensi untuk baris dan satu dimensi lagi untuk kolom. Jadi, MATRIX.dim() atau MATRIX.ndimension() akan mengembalikan nilai 2, menunjukkan bahwa tensor tersebut memiliki dua dimensi.

In [ ]:
MATRIX.shape

torch.Size([2, 2])

Untuk mengetahui jumlah dimensi dari tensor MATRIX yang telah Anda buat, Anda bisa menggunakan metode .dim() atau juga .ndimension() pada tensor PyTorch.

python
Copy code
MATRIX.dim()  # atau MATRIX.ndimension()
Tensor MATRIX yang merupakan matriks 2x2 dalam kasus ini akan memiliki 2 dimensi. Ini karena matriks memiliki struktur dua dimensi: satu dimensi untuk baris dan satu dimensi lagi untuk kolom. Jadi, MATRIX.dim() atau MATRIX.ndimension() akan mengembalikan nilai 2, menunjukkan bahwa tensor tersebut memiliki dua dimensi.

In [ ]:
# Tensor
TENSOR = torch.tensor([[[1, 2, 3],
                        [3, 6, 9],
                        [2, 4, 5]]])
TENSOR

tensor([[[1, 2, 3],
         [3, 6, 9],
         [2, 4, 5]]])

Potongan kode ini membuat variabel TENSOR yang merupakan sebuah tensor PyTorch. Lebih spesifik lagi, ini adalah sebuah tensor tiga dimensi.

Dalam kasus ini, torch.tensor([[[1, 2, 3], [3, 6, 9], [2, 4, 5]]]) menciptakan sebuah tensor dengan struktur tiga dimensi. Tensor ini memiliki bentuk (shape) 1x3x3, yang artinya:

Dimensi pertama memiliki panjang 1 (hanya satu elemen).
Dimensi kedua memiliki panjang 3, yang menunjukkan ada 3 elemen dalam setiap baris.
Dimensi ketiga juga memiliki panjang 3, menandakan ada 3 elemen dalam setiap kolom.
Tensor tiga dimensi seperti ini sering digunakan dalam berbagai algoritma dalam bidang seperti pengolahan citra, dalam model-model deep learning, dan dalam representasi data yang memerlukan struktur tiga dimensi, seperti data volumetrik atau rangkaian waktu spasial. Dalam kasus ini, tensor ini menunjukkan data 3x3 yang terdiri dari nilai-nilai numerik.

In [ ]:
# Check number of dimensions for TENSOR
TENSOR.dim

3

Atribut dim pada tensor PyTorch sebenarnya adalah sebuah metode yang digunakan untuk mendapatkan jumlah dimensi dari tensor. Namun, perlu diingat bahwa dim pada dasarnya adalah sebuah fungsi yang perlu dipanggil dengan tanda kurung "()".

Jika kita ingin mengetahui jumlah dimensi dari tensor TENSOR yang telah Anda buat (yaitu tensor 3D), kita bisa menggunakan metode dim():

python
Copy code
TENSOR.dim()  # atau TENSOR.dim
Pemanggilan ini akan mengembalikan jumlah dimensi dari tensor TENSOR. Dalam kasus tensor TENSOR yang merupakan tensor tiga dimensi, TENSOR.dim() akan menghasilkan nilai 3, menunjukkan bahwa tensor tersebut memiliki tiga dimensi.

In [ ]:
# Check shape of TENSOR
TENSOR.shape

torch.Size([1, 3, 3])

atribut shape pada tensor PyTorch digunakan untuk mengetahui dimensi atau struktur dari tensor. Untuk tensor TENSOR yang telah Anda buat (yaitu tensor 3D), TENSOR.shape akan memberikan informasi tentang jumlah elemen dalam setiap dimensi tensor.

Dalam kasus tensor TENSOR, yang merupakan tensor tiga dimensi, TENSOR.shape akan mengembalikan sebuah tupel yang menunjukkan jumlah elemen dalam setiap dimensi.

python
Copy code
TENSOR.shape
Outputnya akan menjadi (1, 3, 3). Ini menunjukkan bahwa tensor ini memiliki 1 elemen dalam dimensi pertamanya (depth), 3 elemen dalam dimensi kedua (baris), dan 3 elemen dalam dimensi ketiga (kolom)

In [ ]:
# Create a random tensor of size (3, 4)
random_tensor = torch.rand(size=(3, 4))
random_tensor, random_tensor.dtype

(tensor([[0.1484, 0.9199, 0.3102, 0.2975],
         [0.6504, 0.4424, 0.8936, 0.4705],
         [0.8065, 0.0771, 0.3206, 0.3488]]),
 torch.float32)

Potongan kode ini membuat sebuah tensor acak dengan ukuran 3x4 menggunakan PyTorch.

torch.rand(size=(3, 4)) menciptakan tensor berukuran 3x4 yang diisi dengan nilai-nilai acak yang berada di antara 0 dan 1 dari distribusi seragam (uniform distribution).
size=(3, 4) digunakan untuk menentukan ukuran tensor yang ingin Anda buat.
random_tensor akan berisi tensor acak dengan 3 baris dan 4 kolom.
random_tensor.dtype digunakan untuk mendapatkan tipe data (dtype) dari tensor random_tensor. Pada umumnya, tipe data ini akan menjadi float32 secara default karena torch.rand() secara bawaan akan menghasilkan tensor dengan tipe data float32.

Jadi, output dari potongan kode ini adalah random_tensor, yang berisi tensor 3x4 dengan nilai-nilai acak yang memiliki tipe data float32.

In [ ]:
# Create a random tensor of size (224, 224, 3)
random_image_size_tensor = torch.rand(size=(224, 224, 3))
random_image_size_tensor.shape, random_image_size_tensor.ndim

(torch.Size([224, 224, 3]), 3)

Tentunya! Potongan kode ini mencoba membuat tensor acak dengan ukuran (224, 224, 3) menggunakan PyTorch.

Namun, perlu diingat bahwa tensor dalam PyTorch memiliki aturan dimensi yang sedikit berbeda dari tensor dalam format gambar yang biasanya memiliki urutan (tinggi, lebar, channel). Dalam kasus kode ini, tensor yang ingin dibuat memiliki urutan yang tidak umum untuk representasi gambar.

Tensor dalam PyTorch tidak secara langsung mendukung representasi ini karena PyTorch mengikuti konvensi urutan dimensi (dimensi pertama sebagai batch size, dimensi kedua sebagai channel, dan sisanya sebagai tinggi dan lebar). Sehingga, menghasilkan tensor dengan urutan (224, 224, 3) tidak akan menciptakan representasi gambar yang sesuai.

Pesan kesalahan atau peringatan akan muncul karena upaya menciptakan tensor dengan urutan dimensi yang tidak umum dalam PyTorch. Sebaiknya, dalam PyTorch, representasi gambar lebih umum dilakukan dengan urutan dimensi (batch size, channel, tinggi, lebar).

python
Copy code
random_image_size_tensor.shape  # Akan menghasilkan bentuk tensor
random_image_size_tensor.ndim   # Akan menghasilkan jumlah dimensi tensor
Hasilnya akan menunjukkan bahwa upaya membuat tensor dengan urutan dimensi (224, 224, 3) tidak valid dalam konteks tensor PyTorch.

In [ ]:
# Create a tensor of all zeros
zeros = torch.zeros(size=(3, 4))
zeros, zeros.dtype

(tensor([[0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.]]),
 torch.float32)

Kode ini membuat sebuah tensor yang diisi dengan nilai nol dan memiliki ukuran 3x4 menggunakan PyTorch.

torch.zeros(size=(3, 4)) menciptakan sebuah tensor berukuran 3x4 yang diisi dengan nilai-nilai nol.
size=(3, 4) digunakan untuk menentukan ukuran tensor yang ingin Anda buat.
zeros akan berisi tensor dengan 3 baris dan 4 kolom yang semuanya bernilai nol.
zeros.dtype digunakan untuk mendapatkan tipe data (dtype) dari tensor zeros. Secara bawaan, dalam hal ini, tipe data akan menjadi float32 karena torch.zeros() secara default akan menghasilkan tensor dengan tipe data float32.

Jadi, output dari potongan kode ini adalah zeros, yang berisi tensor 3x4 dengan semua nilai yang diisi nol dan memiliki tipe data float32 secara bawaan.

We can do the same to create a tensor of all ones except using [`torch.ones()` ](https://pytorch.org/docs/stable/generated/torch.ones.html) instead.

In [ ]:
# Create a tensor of all ones
ones = torch.ones(size=(3, 4))
ones, ones.dtype

(tensor([[1., 1., 1., 1.],
         [1., 1., 1., 1.],
         [1., 1., 1., 1.]]),
 torch.float32)

Kode ini membuat sebuah tensor yang diisi dengan nilai satu dan memiliki ukuran 3x4 menggunakan PyTorch.

torch.ones(size=(3, 4)) menciptakan sebuah tensor berukuran 3x4 yang diisi dengan nilai-nilai satu.
size=(3, 4) digunakan untuk menentukan ukuran tensor yang ingin Anda buat.
ones akan berisi tensor dengan 3 baris dan 4 kolom yang semua elemennya bernilai satu.
ones.dtype digunakan untuk mendapatkan tipe data (dtype) dari tensor ones. Dalam hal ini, secara bawaan, tipe data akan menjadi float32 karena torch.ones() secara default akan menghasilkan tensor dengan tipe data float32.

Jadi, output dari potongan kode ini adalah ones, yang berisi tensor 3x4 dengan semua nilai yang diisi dengan satu dan memiliki tipe data float32 secara bawaan.

In [ ]:
# Use torch.arange(), torch.range() is deprecated
zero_to_ten_deprecated = torch.range(0, 10) # Note: this may return an error in the future

# Create a range of values 0 to 10
zero_to_ten = torch.arange(start=0, end=10, step=1)
zero_to_ten

<ipython-input-18-a404776195c1>:2: UserWarning: torch.range is deprecated and will be removed in a future release because its behavior is inconsistent with Python's range builtin. Instead, use torch.arange, which produces values in [start, end).
  zero_to_ten_deprecated = torch.range(0, 10) # Note: this may return an error in the future


tensor([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

Potongan kode ini menggunakan fungsi torch.arange() untuk membuat tensor yang berisi urutan nilai dari 0 hingga 9.999 (tidak termasuk 10) dengan langkah 1.

torch.arange(start=0, end=10, step=1) menciptakan tensor yang memuat urutan nilai dari 0 hingga 10 (eksklusif) dengan langkah satu.
start=0 menunjukkan nilai awal dari urutan.
end=10 menunjukkan nilai akhir urutan, namun nilai ini tidak termasuk dalam urutan.
step=1 menunjukkan langkah atau increment antara setiap nilai dalam urutan.
Sehingga, zero_to_ten akan berisi tensor yang memuat nilai-nilai dari 0 hingga 9 dengan inkrement satu, yaitu: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9].

Penggunaan torch.range() pada potongan kode sebelumnya sebenarnya sudah tidak direkomendasikan lagi karena telah di-deprecate, dan rekomendasi yang lebih baik adalah menggunakan torch.arange() seperti yang digunakan dalam kode kedua.

In [ ]:
# Can also create a tensor of zeros similar to another tensor
ten_zeros = torch.zeros_like(input=zero_to_ten) # will have same shape
ten_zeros

tensor([0, 0, 0, 0, 0, 0, 0, 0, 0, 0])

Potongan kode ini menggunakan metode torch.zeros_like() untuk membuat tensor yang berisi nilai-nilai nol dan memiliki bentuk yang sama dengan tensor zero_to_ten.

torch.zeros_like(input=zero_to_ten) menciptakan tensor dengan bentuk yang sama seperti tensor zero_to_ten dan diisi dengan nilai-nilai nol.
input=zero_to_ten menentukan tensor yang digunakan sebagai referensi untuk menentukan bentuk tensor yang baru dibuat (ten_zeros dalam kasus ini).
Sehingga, ten_zeros akan memiliki bentuk yang sama dengan tensor zero_to_ten, yaitu [0, 1, 2, 3, 4, 5, 6, 7, 8, 9], tapi nilainya akan seluruhnya nol. Ini berguna ketika Anda ingin membuat tensor baru dengan bentuk yang sama tetapi diisi dengan nilai tertentu (dalam kasus ini, nol).

In [ ]:
# Default datatype for tensors is float32
float_32_tensor = torch.tensor([3.0, 6.0, 9.0],
                               dtype=None, # defaults to None, which is torch.float32 or whatever datatype is passed
                               device=None, # defaults to None, which uses the default tensor type
                               requires_grad=False) # if True, operations performed on the tensor are recorded

float_32_tensor.shape, float_32_tensor.dtype, float_32_tensor.device

(torch.Size([3]), torch.float32, device(type='cpu'))

kode ini membuat tensor dengan tiga elemen (3.0, 6.0, 9.0) menggunakan PyTorch. Mari kita terangkan argumen-argumen yang digunakan dalam penciptaan tensor ini:

torch.tensor([3.0, 6.0, 9.0], dtype=None, device=None, requires_grad=False) menciptakan tensor dengan nilai-nilai [3.0, 6.0, 9.0].
dtype=None pada kode ini mengikuti default PyTorch yang mengartikan bahwa nilai-nilai dalam tensor akan memiliki tipe data float32 karena itu adalah tipe data default untuk tensor jika tidak ada tipe data yang disediakan.
device=None juga mengikuti default PyTorch, yang berarti tensor akan dibuat pada perangkat yang dipilih sebagai default oleh PyTorch (biasanya CPU, kecuali Anda secara eksplisit memindahkan tensor ke GPU atau perangkat lain).
requires_grad=False menunjukkan bahwa tensor ini tidak memerlukan perhitungan gradien saat melalukan operasi (ini berguna ketika Anda menggunakan tensor dalam konteks pemelajaran mesin dan perhitungan gradien diperlukan untuk pelatihan model).
float_32_tensor.shape akan mengembalikan (3,), menunjukkan bahwa ini adalah tensor dengan satu dimensi yang memiliki tiga elemen.

float_32_tensor.dtype akan mengembalikan torch.float32, menunjukkan bahwa tipe data default float32 digunakan untuk tensor ini.

float_32_tensor.device akan mengembalikan informasi tentang perangkat yang digunakan untuk tensor. Dalam hal ini, karena device=None yang mengikuti default PyTorch, tensor akan berada pada perangkat default yang ditentukan oleh konfigurasi PyTorch.

In [ ]:
float_16_tensor = torch.tensor([3.0, 6.0, 9.0],
                               dtype=torch.float16) # torch.half would also work

float_16_tensor.dtype

torch.float16

Potongan kode ini membuat tensor dengan tiga elemen (3.0, 6.0, 9.0) menggunakan PyTorch, namun kali ini tipe data yang digunakan adalah torch.float16 atau torch.half.

torch.tensor([3.0, 6.0, 9.0], dtype=torch.float16) menciptakan tensor dengan nilai-nilai [3.0, 6.0, 9.0] menggunakan tipe data torch.float16 atau yang biasa disebut sebagai torch.half.
dtype=torch.float16 menentukan bahwa nilai-nilai dalam tensor ini akan memiliki tipe data float16, yang merepresentasikan float dengan presisi setengah (16-bit float).
float_16_tensor.dtype akan mengembalikan torch.float16, menunjukkan bahwa tipe data float16 telah digunakan untuk tensor ini. Ini dapat berguna dalam situasi di mana memori atau kinerja komputasi menjadi kritis, karena float16 menggunakan setengah ukuran memori dari float32, namun bisa mengurangi presisi numerik.

In [ ]:
# Create a tensor
some_tensor = torch.rand(3, 4)

# Find out details about it
print(some_tensor)
print(f"Shape of tensor: {some_tensor.shape}")
print(f"Datatype of tensor: {some_tensor.dtype}")
print(f"Device tensor is stored on: {some_tensor.device}") # will default to CPU

tensor([[0.4932, 0.6637, 0.3758, 0.9709],
        [0.7731, 0.5571, 0.2887, 0.3525],
        [0.1797, 0.3670, 0.1261, 0.4467]])
Shape of tensor: torch.Size([3, 4])
Datatype of tensor: torch.float32
Device tensor is stored on: cpu


Potongan kode ini menggunakan PyTorch untuk membuat sebuah tensor dengan ukuran 3 baris dan 4 kolom, diisi dengan nilai-nilai acak dari distribusi seragam antara 0 dan 1.

Setelah pembuatan tensor, informasi rinci tentang tensor tersebut dicetak. Pertama, isi dari tensor ditampilkan, menunjukkan nilai-nilai acak yang terdapat dalam tensor 3x4. Selanjutnya, informasi tentang bentuk tensor, yang dalam kasus ini adalah (3, 4), menandakan bahwa ini adalah tensor dengan 3 baris dan 4 kolom. Kemudian, tipe data dari tensor, yang secara default adalah float32 karena penggunaan fungsi torch.rand(), diberikan sebagai informasi terkait tensor. Terakhir, perangkat tempat tensor disimpan juga ditampilkan. Jika tidak ada spesifikasi perangkat (seperti penggunaan GPU), PyTorch akan menyimpan tensor pada CPU secara default. Oleh karena itu, output akan menunjukkan perangkat default (biasanya CPU) di mana tensor disimpan, seperti "cpu" atau "cpu:0". Ini memberikan pemahaman mendalam tentang tensor yang telah dibuat menggunakan PyTorch, termasuk nilai, bentuk, tipe data, dan perangkat tempat tensor disimpan.

In [ ]:
# Create a tensor of values and add a number to it
tensor = torch.tensor([1, 2, 3])
tensor + 10

tensor([11, 12, 13])

Potongan kode ini menciptakan sebuah tensor dengan nilai [1, 2, 3] menggunakan PyTorch, kemudian menambahkan nilai 10 ke setiap elemen dalam tensor.

torch.tensor([1, 2, 3]) menciptakan tensor dengan nilai-nilai [1, 2, 3].
tensor + 10 adalah operasi penambahan yang dilakukan pada setiap elemen tensor, yang akan menambahkan nilai 10 ke setiap elemen di dalam tensor.
Hasilnya akan menjadi tensor baru yang memiliki nilai-nilai [11, 12, 13], karena 10 ditambahkan ke setiap elemen yang ada dalam tensor. Operasi ini bertujuan untuk menambahkan nilai skalar (10 dalam hal ini) ke setiap elemen dalam tensor.

In [ ]:
# Multiply it by 10
tensor * 10

tensor([10, 20, 30])

Potongan kode ini melakukan operasi perkalian pada tensor yang telah dibuat sebelumnya.

Setelah membuat tensor dengan nilai [1, 2, 3] menggunakan torch.tensor([1, 2, 3]), kode tensor * 10 mengalikan setiap elemen dalam tensor dengan nilai 10.

Hasilnya adalah tensor baru yang nilai-nilainya adalah [10, 20, 30], di mana setiap elemen dalam tensor awal (yang memiliki nilai [1, 2, 3]) dikalikan dengan 10. Operasi perkalian ini mengalikan setiap elemen dalam tensor dengan nilai skalar (dalam hal ini, 10).

In [ ]:
# Tensors don't change unless reassigned
tensor

tensor([1, 2, 3])

Ketika kita melakukan operasi matematika atau transformasi pada tensor dalam PyTorch, perlu diingat bahwa operasi tersebut tidak secara langsung mengubah nilai-nilai dalam tensor tersebut kecuali jika hasilnya ditugaskan kembali ke tensor yang sama atau tensor baru.

Dalam konteks kode yang Anda tunjukkan, ketika Anda hanya mengetikkan tensor pada baris kode, itu hanya akan menampilkan nilai dari tensor tersebut. Namun, operasi sebelumnya yang dilakukan seperti penambahan atau perkalian (misalnya, tensor + 10 atau tensor * 10) tidak mengubah tensor aslinya kecuali jika hasilnya ditugaskan kembali ke tensor atau pada variabel lain.

Sebagai contoh, jika Anda menulis tensor = tensor * 10 atau new_tensor = tensor * 10, ini akan membuat tensor baru (new_tensor dalam contoh kedua) yang nilainya hasil dari operasi perkalian, sementara tensor aslinya tetap tidak berubah.

Jadi, ketika Anda hanya mengetikkan tensor tanpa operasi pengubahan nilai atau tanpa menugaskan hasilnya ke variabel baru, itu hanya akan menampilkan nilai dari tensor yang sudah ada tanpa melakukan perubahan pada tensor itu sendiri.

In [ ]:
# Subtract and reassign
tensor = tensor - 10
tensor

tensor([-9, -8, -7])

Potongan kode ini melakukan operasi pengurangan dan penugasan kembali pada tensor yang sudah ada sebelumnya.

tensor = tensor - 10: Ini mengurangkan nilai 10 dari setiap elemen dalam tensor yang sudah ada, kemudian menugaskan kembali hasilnya ke variabel tensor.

Operasi ini menghasilkan tensor baru yang nilai-nilainya adalah [0, 10, 20] karena mengurangkan 10 dari setiap elemen dalam tensor asli yang memiliki nilai [10, 20, 30].

Kemudian, ketika tensor ditampilkan setelah operasi ini, hasilnya adalah tensor yang telah diperbarui, yaitu tensor dengan nilai-nilai [0, 10, 20], bukan lagi nilai asli [1, 2, 3]. Operasi ini mengubah nilai-nilai dalam variabel tensor karena hasil operasi pengurangan tersebut ditugaskan kembali ke tensor.

In [ ]:
# Add and reassign
tensor = tensor + 10
tensor

tensor([1, 2, 3])

Potongan kode ini melakukan serangkaian operasi pada sebuah tensor dalam PyTorch.

Pertama, tensor awalnya memiliki nilai [1, 2, 3]. Kemudian, tensor = tensor - 10 mengurangkan nilai 10 dari setiap elemen, menghasilkan tensor baru dengan nilai [0, 10, 20]. Operasi ini mengubah nilai tensor karena hasilnya ditugaskan kembali ke variabel tensor.

Selanjutnya, tensor = tensor + 10 menambahkan nilai 10 ke setiap elemen dalam tensor yang telah diubah sebelumnya menjadi [0, 10, 20]. Hal ini mengakibatkan perubahan nilai menjadi [10, 20, 30]. Operasi ini juga mengubah tensor karena hasilnya ditugaskan kembali ke variabel tensor.

Hasil akhir dari serangkaian operasi ini adalah tensor yang memiliki nilai-nilai baru yaitu [10, 20, 30]. Operasi ini menunjukkan bagaimana operasi matematika pada tensor dalam PyTorch dapat diikuti dengan penugasan kembali untuk memperbarui nilai-nilai dalam tensor.

In [ ]:
# Can also use torch functions
torch.multiply(tensor, 10)

tensor([10, 20, 30])

Tampaknya potongan kode torch.multiply(tensor, 10) mengacu pada upaya menggunakan fungsi torch.multiply() dalam PyTorch untuk mengalikan setiap elemen dalam tensor tensor dengan nilai skalar 10. Namun, dalam PyTorch, tidak ada fungsi bawaan yang disebut torch.multiply(). Untuk melakukan operasi perkalian elemen-wise pada tensor, ada beberapa cara yang dapat dilakukan, seperti menggunakan operator perkalian (*) secara langsung antara tensor dan skalar, atau menggunakan fungsi torch.mul() yang dirancang untuk melakukan operasi perkalian pada tensor. Jika digunakan dengan benar, torch.mul(tensor, 10) akan menghasilkan tensor baru yang berisi nilai-nilai dari setiap elemen dalam tensor dikalikan dengan nilai 10. Ini memberikan kemungkinan untuk mengalikan setiap elemen dalam tensor dengan skalar tertentu, namun, dalam konteks kode yang diberikan, perlu untuk menggunakan cara yang sesuai yang tersedia dalam PyTorch untuk melakukan operasi perkalian elemen-wise pada tensor.

In [ ]:
# Original tensor is still unchanged
tensor

tensor([1, 2, 3])

Ketika Anda melakukan operasi matematika atau transformasi pada sebuah tensor dalam PyTorch, perlu dicatat bahwa operasi tersebut tidak secara langsung mengubah nilai-nilai dalam tensor tersebut kecuali hasilnya ditugaskan kembali ke tensor yang sama atau jika Anda menyimpan hasilnya dalam tensor yang berbeda.

Dalam kasus potongan kode ini, sebelumnya terdapat serangkaian operasi pada tensor tensor, di mana nilai-nilainya telah diubah dari [1, 2, 3] menjadi [10, 20, 30] menggunakan operasi pengurangan dan penambahan. Namun, ketika Anda hanya mengetikkan tensor untuk menampilkan nilainya tanpa melakukan operasi pengubahan atau penugasan hasilnya ke variabel, itu hanya akan menampilkan nilai yang terakhir kali ditugaskan ke tensor.

Hasil yang ditampilkan adalah nilai terakhir dari tensor setelah dilakukan operasi pengurangan dan penambahan. Meskipun nilai-nilai tensor telah diubah sebelumnya, namun jika hasilnya tidak ditugaskan kembali ke tensor atau disimpan dalam variabel lain, nilai-nilai tersebut tidak akan terbarui dan tensor akan tetap menampilkan nilai terakhir yang telah ditugaskan padanya. Operasi yang dilakukan tidak mempengaruhi tensor secara langsung kecuali hasilnya ditugaskan kembali.

In [ ]:
# Element-wise multiplication (each element multiplies its equivalent, index 0->0, 1->1, 2->2)
print(tensor, "*", tensor)
print("Equals:", tensor * tensor)

tensor([1, 2, 3]) * tensor([1, 2, 3])
Equals: tensor([1, 4, 9])


Potongan kode ini berfokus pada operasi perkalian elemen-wise (perkalian satu-satu) dari tensor dengan dirinya sendiri, yaitu setiap elemen dalam tensor akan dikalikan dengan elemen yang memiliki indeks yang sama dalam tensor yang sama.

Pertama, print(tensor, "*", tensor) mencetak nilai tensor dan tanda "*" secara berurutan untuk menunjukkan bahwa operasi perkalian elemen-wise akan dilakukan pada tensor tersebut dengan dirinya sendiri.

Kemudian, print("Equals:", tensor * tensor) mencetak hasil dari operasi perkalian elemen-wise dari tensor dengan dirinya sendiri. Operasi ini menghasilkan tensor baru yang berisi hasil perkalian dari setiap elemen dalam tensor dengan elemen yang memiliki indeks yang sama dalam tensor yang sama.

Sebagai contoh, jika nilai awal tensor adalah [10, 20, 30], hasil perkalian elemen-wise akan menjadi [100, 400, 900]. Hal ini terjadi karena setiap elemen dalam tensor ([10, 20, 30]) dikalikan dengan elemen yang memiliki indeks yang sama dalam tensor yang sama, sehingga [10 * 10, 20 * 20, 30 * 30] menghasilkan [100, 400, 900].

Jadi, operasi tensor * tensor pada dasarnya melakukan perkalian elemen-wise antara tensor dan dirinya sendiri, menghasilkan tensor baru yang berisi hasil perkalian dari setiap elemen dalam tensor dengan elemen yang memiliki indeks yang sama dalam tensor yang sama.

In [ ]:
import torch
tensor = torch.tensor([1, 2, 3])
tensor.shape

torch.Size([3])

Potongan kode ini menggunakan PyTorch untuk membuat sebuah tensor dengan nilai [1, 2, 3]. Setelah tensor tersebut dibuat, dilakukan operasi untuk mendapatkan informasi tentang bentuk (shape) dari tensor tersebut.

torch.tensor([1, 2, 3]) menciptakan tensor dengan nilai-nilai [1, 2, 3]. Tensor ini adalah tensor satu dimensi atau sering disebut sebagai vektor.

Kemudian, tensor.shape digunakan untuk mengakses properti bentuk (shape) dari tensor yang telah dibuat sebelumnya. Bentuk (shape) dari tensor menunjukkan jumlah elemen dalam setiap dimensi. Dalam kasus ini, karena tensor yang dibuat adalah tensor satu dimensi, properti shape akan mengembalikan tupel yang menyatakan jumlah elemen dalam dimensi tunggal tersebut. Dalam kasus ini, hasilnya akan menjadi (3,), menunjukkan bahwa tensor memiliki 3 elemen dalam dimensi tunggalnya. Artinya, ini adalah tensor dengan satu dimensi yang berisi 3 elemen.

In [ ]:
# Element-wise matrix multiplication
tensor * tensor

tensor([1, 4, 9])

Potongan kode tensor * tensor pada dasarnya melakukan operasi perkalian elemen-wise antara tensor dan dirinya sendiri.

Jika tensor memiliki dimensi yang sesuai untuk operasi perkalian elemen-wise, maka setiap elemen dalam tensor akan dikalikan dengan elemen yang memiliki indeks yang sama dalam tensor yang sama. Operasi ini tidak menjalankan perkalian matriks atau dot product, melainkan melakukan perkalian satu-satu pada elemen yang berada pada posisi yang sama dalam tensor.

Contohnya, jika kita memiliki tensor awal dengan nilai [1, 2, 3], hasil dari operasi tensor * tensor akan menjadi [1 * 1, 2 * 2, 3 * 3], yaitu [1, 4, 9]. Ini berarti setiap elemen dalam tensor akan dikalikan dengan elemen yang memiliki indeks yang sama dalam tensor yang sama, menghasilkan tensor baru yang berisi hasil perkalian elemen-wise tersebut.

Namun, perlu diingat bahwa operasi ini berbeda dari operasi perkalian matriks atau dot product. Untuk melakukan perkalian matriks dalam PyTorch, Anda dapat menggunakan fungsi seperti torch.matmul() atau @ operator pada tensor-tensor yang memenuhi persyaratan untuk operasi perkalian matriks. Operasi ini berbeda dan lebih terkait dengan hubungan antara dimensi dari tensor yang bersangkutan.

In [ ]:
# Matrix multiplication
torch.matmul(tensor, tensor)

tensor(14)

Potongan kode torch.matmul(tensor, tensor) dalam PyTorch merupakan panggilan untuk melakukan operasi perkalian matriks antara dua tensor.

Dalam konteks ini, tensor yang diberikan adalah tensor satu dimensi atau vektor dengan nilai [1, 2, 3]. Dalam operasi perkalian matriks, kedua tensor harus memiliki dimensi yang sesuai untuk operasi tersebut dilakukan dengan benar. Dalam kasus ini, karena kedua tensor memiliki dimensi yang sesuai untuk operasi perkalian matriks (tensor satu dimensi), operasi torch.matmul() akan melakukan perkalian matriks antara tensor dengan dirinya sendiri.

Namun, operasi ini bukan perkalian elemen-wise seperti yang dilakukan dengan tensor * tensor. Operasi torch.matmul() akan menghasilkan sebuah skalar, bukan tensor, karena kedua tensor yang diberikan adalah vektor atau tensor satu dimensi.

Hasilnya adalah skalar yang merupakan hasil dari perkalian matriks antara vektor [1, 2, 3] dengan dirinya sendiri. Untuk vektor, hasil perkalian matriks dari vektor dengan dirinya sendiri diperoleh dengan melakukan dot product atau perkalian titik antara vektor tersebut, yaitu (1 * 1) + (2 * 2) + (3 * 3), yang akan menghasilkan nilai 14. Jadi, output dari operasi torch.matmul(tensor, tensor) dalam konteks ini adalah skalar dengan nilai 14.

In [ ]:
# Can also use the "@" symbol for matrix multiplication, though not recommended
tensor @ tensor

tensor(14)

Potongan kode tensor @ tensor menggunakan operator @ untuk melakukan operasi perkalian matriks antara dua tensor dalam Python dengan PyTorch.

Dalam konteks ini, tensor adalah tensor satu dimensi atau vektor dengan nilai [1, 2, 3]. Penggunaan operator @ untuk melakukan perkalian matriks antara dua tensor memiliki fungsi yang sama dengan torch.matmul(), namun dapat ditulis dengan sintaks yang lebih ringkas.

Namun, penggunaan operator @ untuk operasi perkalian matriks dalam PyTorch tidak selalu direkomendasikan karena dapat menjadi ambigu terutama dalam konteks kode yang lebih kompleks atau ketika menggunakan library lain yang juga menggunakan operator @ untuk tujuan yang berbeda. Meskipun demikian, secara fungsional, tensor @ tensor akan melakukan perkalian matriks antara vektor [1, 2, 3] dengan dirinya sendiri, dan hasilnya akan sama dengan yang diperoleh dari torch.matmul(tensor, tensor).

Hasilnya juga akan menjadi skalar dengan nilai 14, karena perkalian matriks dari vektor dengan dirinya sendiri pada dasarnya melakukan dot product atau perkalian titik antara vektor tersebut, yang menghasilkan nilai yang sama seperti yang dijelaskan sebelumnya.

In [ ]:
%%time
# Matrix multiplication by hand
# (avoid doing operations with for loops at all cost, they are computationally expensive)
value = 0
for i in range(len(tensor)):
  value += tensor[i] * tensor[i]
value

CPU times: user 1.35 ms, sys: 0 ns, total: 1.35 ms
Wall time: 6.42 ms


tensor(14)

Potongan kode ini mencoba melakukan operasi perkalian matriks (dot product) secara manual pada vektor atau tensor satu dimensi yang disimpan dalam variabel tensor.

Dalam konteks ini, kode menggunakan iterasi for loop untuk mengakses setiap elemen dalam vektor tensor, dan kemudian mengalikan setiap elemen dengan dirinya sendiri, menambahkan hasil perkalian tersebut ke variabel value.

Namun, penting untuk dicatat bahwa metode ini tidak melakukan operasi perkalian matriks secara eksplisit. Sebaliknya, ini melakukan operasi dot product atau perkalian titik antara vektor tensor dengan dirinya sendiri secara manual, di mana setiap elemen dalam vektor dikalikan dengan elemen yang memiliki indeks yang sama dalam vektor yang sama, dan hasilnya dijumlahkan dalam variabel value.

Meskipun ini memberikan hasil yang benar untuk dot product pada vektor atau tensor satu dimensi, namun menggunakan loop for secara langsung untuk operasi ini tidak dianjurkan, terutama untuk ukuran data yang besar, karena operasi ini lebih lambat secara komputasi daripada menggunakan operasi matriks yang dioptimalkan yang disediakan oleh fungsi PyTorch seperti torch.matmul() atau operator @. Operasi dengan loop for dapat menjadi sangat lambat untuk ukuran data yang besar karena memakan waktu komputasi yang lebih tinggi. Dalam konteks ini, kode ini bertujuan untuk menunjukkan konsep operasi dot product secara manual, namun tidak optimal dalam hal kinerja.

In [ ]:
%%time
torch.matmul(tensor, tensor)

CPU times: user 36 µs, sys: 0 ns, total: 36 µs
Wall time: 39.6 µs


tensor(14)

Potongan kode ini menggunakan fungsi PyTorch torch.matmul() untuk melakukan operasi perkalian matriks antara dua tensor.

Dalam konteks ini, tensor adalah tensor satu dimensi atau vektor dengan nilai [1, 2, 3]. Penggunaan torch.matmul() pada vektor dengan dirinya sendiri akan menghasilkan hasil perkalian matriks dari vektor tersebut.

Penggunaan fungsi ini merupakan cara yang lebih efisien secara komputasi daripada menggunakan pendekatan manual dengan loop for, terutama untuk operasi perkalian matriks pada ukuran data yang besar. Fungsi ini diimplementasikan secara efisien di tingkat yang lebih rendah menggunakan optimisasi yang dioptimalkan, sehingga memberikan kinerja yang lebih baik.

Penggunaan %%time sebelum perintah dijalankan digunakan untuk mengukur waktu yang diperlukan untuk mengeksekusi sel selanjutnya dalam notebook atau lingkungan yang mendukung magic commands seperti Jupyter Notebook atau Google Colab. Ini membantu untuk mengukur waktu eksekusi suatu perintah atau blok kode, dan memberikan perkiraan waktu yang dibutuhkan untuk menyelesaikan perintah tersebut. Dalam hal ini, ini digunakan untuk mengukur waktu yang dibutuhkan oleh operasi torch.matmul(tensor, tensor) untuk dijalankan dan menampilkan hasilnya dalam waktu yang terukur.

In [ ]:
# Shapes need to be in the right way
tensor_A = torch.tensor([[1, 2],
                         [3, 4],
                         [5, 6]], dtype=torch.float32)

tensor_B = torch.tensor([[7, 10],
                         [8, 11],
                         [9, 12]], dtype=torch.float32)

torch.matmul(tensor_A, tensor_B) # (this will error)

RuntimeError: ignored

Potongan kode ini mencoba melakukan operasi perkalian matriks antara dua tensor, tensor_A dan tensor_B, menggunakan fungsi torch.matmul() dalam PyTorch.

Namun, operasi ini akan menghasilkan error karena aturan dimensi yang tidak memenuhi syarat untuk operasi perkalian matriks yang valid. Dalam operasi perkalian matriks, dimensi dari tensor-tensor yang dilibatkan harus sesuai untuk operasi tersebut.

tensor_A adalah sebuah tensor dengan bentuk (shape) 3x2, dan tensor_B adalah tensor dengan bentuk (shape) 3x2. Dalam operasi perkalian matriks yang valid, aturan dimensi yang berlaku adalah jumlah kolom dalam tensor pertama harus sama dengan jumlah baris dalam tensor kedua. Dalam kasus ini, tensor pertama (tensor_A) memiliki 2 kolom, sedangkan tensor kedua (tensor_B) memiliki 3 baris.

Oleh karena itu, tidak mungkin melakukan operasi perkalian matriks antara tensor_A dan tensor_B karena ketidaksesuaian dimensi, dan hal ini akan menyebabkan error. Operasi perkalian matriks hanya dapat dilakukan jika dimensi dari kedua tensor memenuhi aturan yang diperlukan, yaitu jumlah kolom dalam tensor pertama harus sama dengan jumlah baris dalam tensor kedua untuk mendapatkan hasil yang valid.

In [ ]:
import torch

tensor_A = torch.tensor([[1, 2],
                         [3, 4],
                         [5, 6]], dtype=torch.float32)

tensor_B = torch.tensor([[7, 10],
                         [8, 11],
                         [9, 12]], dtype=torch.float32)

result = torch.matmul(tensor_A, tensor_B.t()) # Transposing tensor_B

print(result)


tensor([[ 27.,  30.,  33.],
        [ 61.,  68.,  75.],
        [ 95., 106., 117.]])


Potongan kode ini melakukan operasi perkalian matriks antara dua tensor, tensor_A dan tensor_B, menggunakan fungsi torch.matmul() dalam PyTorch. Namun, sebelumnya, tensor_B telah ditranspose sebelum operasi perkalian matriks dilakukan.

tensor_A adalah tensor dengan bentuk (shape) 3x2, dan tensor_B adalah tensor dengan bentuk (shape) 3x2. Dalam kasus ini, tensor_B ditranspose menggunakan metode .t(), yang mengubah baris menjadi kolom dan sebaliknya, sehingga tensor_B.t() memiliki bentuk (shape) 2x3.

Setelah melakukan operasi transpose pada tensor_B, torch.matmul(tensor_A, tensor_B.t()) akan melakukan operasi perkalian matriks antara tensor_A dan tensor_B yang telah ditranspose. Dalam aturan perkalian matriks, jumlah kolom dalam tensor pertama (tensor_A) harus sama dengan jumlah baris dalam tensor kedua (tensor_B.t()). Dalam kasus ini, tensor_A memiliki 2 kolom dan tensor_B.t() memiliki 2 baris setelah dilakukan operasi transpose, sehingga aturan dimensi perkalian matriks terpenuhi.

Hasil operasi perkalian matriks ini akan disimpan dalam variabel result, yang akan berisi hasil dari perkalian matriks antara tensor_A dan tensor_B yang telah ditranspose. Kemudian, hasil tersebut akan dicetak menggunakan perintah print(result).

In [ ]:
# View tensor_A and tensor_B
print(tensor_A)
print(tensor_B)

tensor([[1., 2.],
        [3., 4.],
        [5., 6.]])
tensor([[ 7., 10.],
        [ 8., 11.],
        [ 9., 12.]])


kode ini menggunakan perintah print() untuk menampilkan nilai dari dua tensor yang telah didefinisikan sebelumnya, yaitu tensor_A dan tensor_B. Tensor tensor_A memiliki bentuk (shape) 3x2, yang berarti terdiri dari tiga baris dan dua kolom. Saat ditampilkan dengan print(tensor_A), hasilnya akan menunjukkan nilai-nilai tensor dalam format yang terstruktur, menampilkan setiap elemen sesuai dengan posisi baris dan kolomnya. Pada contoh ini, nilai-nilai dari tensor_A adalah:

lua
Copy code
tensor([[1., 2.],
        [3., 4.],
        [5., 6.]])
Sementara itu, tensor tensor_B juga memiliki bentuk (shape) 3x2, sehingga terdiri dari tiga baris dan dua kolom yang sama seperti tensor_A. Ketika di-print menggunakan print(tensor_B), hasilnya menampilkan nilai-nilai dalam tensor ini, menunjukkan elemen-elemen tensor yang memiliki nilai yang berbeda dari tensor_A. Contoh nilai-nilai dari tensor_B adalah:

lua
Copy code
tensor([[ 7., 10.],
        [ 8., 11.],
        [ 9., 12.]])
Kedua tensor tersebut memiliki bentuk yang serupa tetapi berbeda nilai-nilainya, yang tercermin dalam hasil cetakan ketika mereka ditampilkan. Ini memperlihatkan isi atau nilai-nilai dari masing-masing tensor dalam format yang mudah dibaca, membantu dalam memeriksa dan memahami nilai-nilai yang disimpan dalam tensor.

In [ ]:
# View tensor_A and tensor_B.T
print(tensor_A)
print(tensor_B.T)

tensor([[1., 2.],
        [3., 4.],
        [5., 6.]])
tensor([[ 7.,  8.,  9.],
        [10., 11., 12.]])


Kode ini menggunakan perintah print() untuk menampilkan nilai dari dua tensor yang telah didefinisikan sebelumnya, yaitu tensor_A dan tensor_B yang sudah ditranspose (tensor_B.T).

print(tensor_A) menampilkan nilai dari tensor tensor_A, yang memiliki bentuk (shape) 3x2, atau tiga baris dan dua kolom. Nilai-nilai tensor ini akan ditampilkan dalam format terstruktur yang menunjukkan setiap elemen sesuai dengan posisi baris dan kolomnya. Contoh nilai-nilai dari tensor_A adalah:
lua
Copy code
tensor([[1., 2.],
        [3., 4.],
        [5., 6.]])
print(tensor_B.T) menampilkan nilai dari tensor tensor_B yang telah ditranspose menggunakan metode .T(). Transpose mengubah baris menjadi kolom dan sebaliknya. Dalam konteks ini, tensor tensor_B memiliki bentuk (shape) 3x2, dan setelah dilakukan transpose, bentuknya menjadi 2x3 (dua baris dan tiga kolom). Ketika di-print, tensor_B.T akan menampilkan nilai-nilai tensor yang telah ditranspose ini, menunjukkan elemen-elemen tensor yang disusun ulang dalam format 2x3. Contoh nilai-nilai dari tensor_B.T adalah:
lua
Copy code
tensor([[ 7.,  8.,  9.],
        [10., 11., 12.]])
Perbedaan antara tensor_B dan tensor_B.T adalah susunan nilai-nilai tensor dalam format yang berbeda karena transpose mengubah struktur baris dan kolom, sehingga tensor_B.T menampilkan tensor dengan baris dan kolom yang telah ditukar posisinya dari tensor_B.

In [ ]:
# The operation works when tensor_B is transposed
print(f"Original shapes: tensor_A = {tensor_A.shape}, tensor_B = {tensor_B.shape}\n")
print(f"New shapes: tensor_A = {tensor_A.shape} (same as above), tensor_B.T = {tensor_B.T.shape}\n")
print(f"Multiplying: {tensor_A.shape} * {tensor_B.T.shape} <- inner dimensions match\n")
print("Output:\n")
output = torch.matmul(tensor_A, tensor_B.T)
print(output)
print(f"\nOutput shape: {output.shape}")

Original shapes: tensor_A = torch.Size([3, 2]), tensor_B = torch.Size([3, 2])

New shapes: tensor_A = torch.Size([3, 2]) (same as above), tensor_B.T = torch.Size([2, 3])

Multiplying: torch.Size([3, 2]) * torch.Size([2, 3]) <- inner dimensions match

Output:

tensor([[ 27.,  30.,  33.],
        [ 61.,  68.,  75.],
        [ 95., 106., 117.]])

Output shape: torch.Size([3, 3])


ode ini memperlihatkan operasi perkalian matriks antara dua tensor, tensor_A dan tensor_B yang sudah ditranspose (tensor_B.T), serta menampilkan informasi tentang bentuk (shape) dari tensor sebelum dan setelah dilakukan transpose.

print(f"Original shapes: tensor_A = {tensor_A.shape}, tensor_B = {tensor_B.shape}\n") menampilkan bentuk (shape) awal dari tensor tensor_A dan tensor_B. Pada contoh ini, tensor_A adalah tensor dengan bentuk 3x2 (tiga baris dan dua kolom), sedangkan tensor_B juga memiliki bentuk yang sama, yaitu 3x2.

print(f"New shapes: tensor_A = {tensor_A.shape} (same as above), tensor_B.T = {tensor_B.T.shape}\n") menampilkan bentuk (shape) baru setelah melakukan transpose pada tensor_B. Setelah dilakukan transpose, tensor_B.T memiliki bentuk 2x3 (dua baris dan tiga kolom), sementara bentuk tensor_A tetap 3x2.

print(f"Multiplying: {tensor_A.shape} * {tensor_B.T.shape} <- inner dimensions match\n") memberikan informasi tentang dimensi tensor yang akan dikalikan. Inner dimensions (dimensi dalam) dari kedua tensor adalah 3 dan 2, yang memenuhi persyaratan operasi perkalian matriks karena jumlah kolom dari tensor_A sama dengan jumlah baris dari tensor_B.T.

output = torch.matmul(tensor_A, tensor_B.T) melakukan operasi perkalian matriks antara tensor_A dan tensor_B.T. Karena kedua tensor memiliki dimensi yang sesuai untuk operasi perkalian matriks, hasilnya adalah tensor baru yang berisi hasil perkalian matriks antara kedua tensor tersebut.

print(output) menampilkan hasil dari operasi perkalian matriks antara tensor_A dan tensor_B.T.

print(f"\nOutput shape: {output.shape}") menampilkan bentuk (shape) dari hasil operasi, yang dalam kasus ini adalah tensor dengan bentuk 3x3 (tiga baris dan tiga kolom), karena hasil perkalian matriks antara tensor_A (3x2) dan tensor_B.T (2x3) akan menghasilkan tensor dengan bentuk 3x3.

You can also use [`torch.mm()`](https://pytorch.org/docs/stable/generated/torch.mm.html) which is a short for `torch.matmul()`.

In [ ]:
# torch.mm is a shortcut for matmul
torch.mm(tensor_A, tensor_B.T)

tensor([[ 27.,  30.,  33.],
        [ 61.,  68.,  75.],
        [ 95., 106., 117.]])

In [ ]:
# Since the linear layer starts with a random weights matrix, let's make it reproducible (more on this later)
torch.manual_seed(42)
# This uses matrix multiplication
linear = torch.nn.Linear(in_features=2, # in_features = matches inner dimension of input
                         out_features=6) # out_features = describes outer value
x = tensor_A
output = linear(x)
print(f"Input shape: {x.shape}\n")
print(f"Output:\n{output}\n\nOutput shape: {output.shape}")

Input shape: torch.Size([3, 2])

Output:
tensor([[2.2368, 1.2292, 0.4714, 0.3864, 0.1309, 0.9838],
        [4.4919, 2.1970, 0.4469, 0.5285, 0.3401, 2.4777],
        [6.7469, 3.1648, 0.4224, 0.6705, 0.5493, 3.9716]],
       grad_fn=<AddmmBackward0>)

Output shape: torch.Size([3, 6])


In [ ]:
# Create a tensor
x = torch.arange(0, 100, 10)
x

tensor([ 0, 10, 20, 30, 40, 50, 60, 70, 80, 90])

Potongan kode ini menggunakan fungsi torch.arange() untuk membuat tensor x. Fungsi ini menghasilkan urutan nilai berdasarkan parameter yang diberikan. Dalam kasus ini, torch.arange(0, 100, 10) membuat tensor dengan nilai mulai dari 0 hingga kurang dari 100, dengan interval 10 antara setiap nilai. Ini berarti tensor x akan berisi nilai dari 0 hingga 90 dengan selisih 10 antara setiap nilai berturut-turut.

Hasilnya adalah tensor satu dimensi atau sering disebut sebagai vektor dengan nilai-nilai sebagai berikut:

scss
Copy code
tensor([ 0, 10, 20, 30, 40, 50, 60, 70, 80, 90])
Setiap elemen dalam tensor ini adalah hasil dari nilai-nilai yang di-generate oleh torch.arange() dengan menggunakan parameter yang diberikan, yaitu dimulai dari 0, kurang dari 100, dan dengan selisih 10 antara setiap nilai berturut-turut. Operasi ini menghasilkan urutan nilai yang bertambah secara berurutan sesuai dengan interval yang ditentukan.

Now let's perform some aggregation.

In [ ]:
print(f"Minimum: {x.min()}")
print(f"Maximum: {x.max()}")
# print(f"Mean: {x.mean()}") # this will error
print(f"Mean: {x.type(torch.float32).mean()}") # won't work without float datatype
print(f"Sum: {x.sum()}")

Minimum: 0
Maximum: 90
Mean: 45.0
Sum: 450


Potongan kode ini berfokus pada operasi statistik yang dilakukan pada tensor x, yang telah dibuat sebelumnya menggunakan torch.arange().

print(f"Minimum: {x.min()}") menampilkan nilai minimum dari tensor x dengan menggunakan metode .min(). Ini mengembalikan nilai minimum dari seluruh elemen dalam tensor x, yang dalam hal ini adalah 0.

print(f"Maximum: {x.max()}") menampilkan nilai maksimum dari tensor x dengan menggunakan metode .max(). Ini mengembalikan nilai maksimum dari seluruh elemen dalam tensor x, yang dalam hal ini adalah 90.

print(f"Mean: {x.type(torch.float32).mean()}") berusaha untuk menghitung nilai rata-rata dari tensor x menggunakan metode .mean(). Namun, perlu diketahui bahwa operasi .mean() tidak dapat langsung diterapkan pada tensor yang memiliki tipe data non-floating point (non-float). Sebagai solusi, kode ini melakukan konversi tipe data tensor x ke float menggunakan x.type(torch.float32) sebelum memanggil .mean(). Hal ini memungkinkan perhitungan nilai rata-rata dari tensor x, yang dalam hal ini adalah 45.0.

print(f"Sum: {x.sum()}") menghitung jumlah dari seluruh elemen dalam tensor x menggunakan metode .sum(). Hasilnya adalah penjumlahan dari seluruh nilai dalam tensor x, yang dalam hal ini adalah 450.

Penting untuk diperhatikan bahwa perhitungan nilai rata-rata (mean()) tidak dapat langsung dilakukan pada tensor yang memiliki tipe data non-floating point, sehingga konversi tipe data ke float diperlukan agar operasi tersebut dapat berhasil dieksekusi.








In [ ]:
torch.max(x), torch.min(x), torch.mean(x.type(torch.float32)), torch.sum(x)

(tensor(90), tensor(0), tensor(45.), tensor(450))

In [ ]:
# Create a tensor
tensor = torch.arange(10, 100, 10)
print(f"Tensor: {tensor}")

# Returns index of max and min values
print(f"Index where max value occurs: {tensor.argmax()}")
print(f"Index where min value occurs: {tensor.argmin()}")

Tensor: tensor([10, 20, 30, 40, 50, 60, 70, 80, 90])
Index where max value occurs: 8
Index where min value occurs: 0


Potongan kode ini menggabungkan beberapa operasi PyTorch pada tensor x, yang telah dibuat sebelumnya menggunakan torch.arange().

torch.max(x) digunakan untuk mencari nilai maksimum dari tensor x. Fungsi torch.max() mengembalikan nilai maksimum dari tensor, bersama dengan indeks dari nilai maksimum tersebut. Dalam kasus ini, perintah ini hanya mengembalikan nilai maksimum dari tensor x, yaitu 90.

torch.min(x) digunakan untuk menemukan nilai minimum dari tensor x. Sama seperti torch.max(), torch.min() mengembalikan nilai minimum dari tensor, bersama dengan indeks dari nilai minimum tersebut. Dalam hal ini, perintah ini hanya mengembalikan nilai minimum dari tensor x, yaitu 0.

torch.mean(x.type(torch.float32)) menghitung nilai rata-rata dari tensor x. Operasi ini pertama-tama mengubah tipe data tensor x ke float menggunakan x.type(torch.float32) agar operasi mean() dapat diterapkan pada tensor dengan tipe data floating point. Kemudian, perintah ini menghitung rata-rata dari seluruh elemen dalam tensor x, yang dalam hal ini adalah 45.0.

torch.sum(x) menghitung jumlah dari seluruh elemen dalam tensor x. Operasi ini akan menjumlahkan semua nilai yang ada dalam tensor x dan mengembalikan hasilnya, yang dalam hal ini adalah 450.

Kombinasi dari perintah-perintah tersebut memberikan informasi statistik umum mengenai tensor x, seperti nilai maksimum, nilai minimum, nilai rata-rata, dan jumlah dari seluruh elemen yang ada dalam tensor tersebut.

In [ ]:
# Create a tensor and check its datatype
tensor = torch.arange(10., 100., 10.)
tensor.dtype

torch.float32

Potongan kode ini berfungsi untuk membuat sebuah tensor menggunakan fungsi torch.arange(), kemudian mengecek tipe data (datatype) dari tensor yang telah dibuat.

torch.arange(10., 100., 10.) digunakan untuk membuat tensor yang berisi urutan nilai, dimulai dari 10, berakhir sebelum 100, dengan selisih antara nilai-nilai berturut-turut adalah 10. Dalam hal ini, angka 10.0 merupakan nilai awal dari urutan, 100.0 adalah nilai yang berhenti sebelum mencapai 100.0, dan 10.0 adalah selisih antara setiap nilai berturut-turut dalam urutan tersebut.

tensor.dtype memeriksa dan mengembalikan tipe data (datatype) dari tensor yang telah dibuat sebelumnya. Kode tersebut akan menghasilkan tipe data dari tensor, yaitu torch.float32 (atau float32). Hal ini karena nilai-nilai dalam tensor dibuat dengan tipe data float (nilai desimal), dan secara default, jika tipe data tidak ditentukan secara eksplisit, PyTorch akan menggunakan torch.float32 sebagai tipe data bawaan untuk tensor numerik.

Dalam konteks ini, kode tersebut menghasilkan dan memeriksa tensor yang berisi urutan nilai-nilai dalam format floating point (desimal) dengan tipe data float32.

Now we'll create another tensor the same as before but change its datatype to `torch.float16`.



In [ ]:
# Create a float16 tensor
tensor_float16 = tensor.type(torch.float16)
tensor_float16

tensor([10., 20., 30., 40., 50., 60., 70., 80., 90.], dtype=torch.float16)

Potongan kode ini mengubah tipe data (datatype) dari sebuah tensor yang telah dibuat sebelumnya, tensor, menjadi torch.float16 menggunakan metode .type() dari PyTorch.

tensor.type(torch.float16) adalah operasi yang mengonversi tensor yang sudah ada (tensor) menjadi tipe data torch.float16, yang merupakan tipe data floating point dengan presisi setengah (16-bit). Operasi ini menghasilkan tensor baru dengan tipe data yang diubah sesuai dengan yang diinginkan.

tensor_float16 merupakan hasil dari operasi konversi tersebut, dan akan berisi nilai-nilai yang sama dengan tensor yang asli, tetapi dalam tipe data torch.float16. Penting untuk dicatat bahwa penggunaan tipe data float16 dapat mengurangi presisi dari nilai-nilai dalam tensor, namun pada saat yang sama, dapat menghemat ruang memori dan mempercepat operasi komputasi, terutama saat bekerja dengan perangkat keras yang mendukung operasi floating point setengah (16-bit).

Dalam konteks ini, tensor_float16 akan berisi nilai-nilai yang sama dengan tensor, tetapi dengan tipe data yang berbeda, yaitu torch.float16. Penyesuaian tipe data tensor sering kali digunakan untuk mengoptimalkan penggunaan memori dan kinerja komputasi pada berbagai operasi yang membutuhkan presisi yang lebih rendah.

And we can do something similar to make a `torch.int8` tensor.

In [ ]:
# Create a int8 tensor
tensor_int8 = tensor.type(torch.int8)
tensor_int8

tensor([10, 20, 30, 40, 50, 60, 70, 80, 90], dtype=torch.int8)


Potongan kode ini mencoba untuk mengubah tipe data (datatype) dari sebuah tensor yang telah dibuat sebelumnya, tensor, menjadi torch.int8 menggunakan metode .type() dari PyTorch.

Namun, terdapat sebuah perhatian penting terkait dengan konversi ke torch.int8. Tipe data int8 merupakan tipe data integer dengan 8-bit. Rentang nilai yang dapat direpresentasikan dalam tipe data ini adalah dari -128 hingga 127. Namun, urutan nilai yang dihasilkan oleh torch.arange(10., 100., 10.) tidak dapat direpresentasikan dalam tipe data int8 karena nilai terkecil dalam urutan tersebut (nilai 10.0) lebih besar daripada nilai minimum yang dapat direpresentasikan dalam int8 (yaitu -128). Akibatnya, konversi tersebut tidak dapat dilakukan secara tepat.

Kode yang diberikan (tensor.type(torch.int8)) pada dasarnya berusaha untuk mengubah tipe data tensor tensor menjadi torch.int8. Namun, karena batasan representasi nilai pada int8, hasilnya akan terbatas pada nilai-nilai yang dapat direpresentasikan dalam tipe data int8, dan nilai-nilai di luar rentang tersebut akan terpotong atau diubah menjadi nilai yang dapat direpresentasikan.

Penting untuk dicatat bahwa operasi konversi ke int8 memerlukan perhatian khusus terhadap rentang nilai yang dapat direpresentasikan dalam tipe data tersebut, dan memastikan bahwa nilai-nilai yang ada tidak melebihi batas yang diperbolehkan dalam tipe data int8 untuk menghindari potensi kehilangan informasi atau perubahan nilai yang tidak diinginkan.

In [ ]:
# Create a tensor
import torch
x = torch.arange(1., 8.)
x, x.shape

(tensor([1., 2., 3., 4., 5., 6., 7.]), torch.Size([7]))


Potongan kode ini menggunakan PyTorch untuk membuat sebuah tensor bernama x menggunakan fungsi torch.arange().

torch.arange(1., 8.) menciptakan sebuah tensor yang berisi urutan nilai dari 1.0 hingga kurang dari 8.0 dengan selisih 1 antara setiap nilai berturut-turut. Dalam hal ini, nilai 1.0 adalah nilai awal dari urutan, dan 8.0 adalah nilai yang berhenti sebelum mencapai 8.0.

x merupakan hasil dari operasi torch.arange(), dan akan berisi nilai-nilai dalam urutan tersebut. Dalam hal ini, x akan berisi nilai-nilai berikut: tensor ([1., 2., 3., 4., 5., 6., 7.]).

x.shape adalah atribut yang digunakan untuk memeriksa bentuk (shape) dari tensor x. Hasilnya adalah (7,), yang menunjukkan bahwa x adalah tensor satu dimensi (atau sering disebut vektor) dengan panjang 7 (memiliki 7 elemen).

Jadi, secara keseluruhan, kode tersebut membuat tensor x yang berisi urutan nilai dari 1.0 hingga 7.0 dengan selisih 1 antara setiap nilai berturut-turut, dan x merupakan vektor satu dimensi dengan panjang 7.







Now let's add an extra dimension with `torch.reshape()`.

In [ ]:
# Add an extra dimension
x_reshaped = x.reshape(1, 7)
x_reshaped, x_reshaped.shape

(tensor([[1., 2., 3., 4., 5., 6., 7.]]), torch.Size([1, 7]))

Potongan kode ini menggunakan metode .reshape() dari PyTorch untuk mengubah tampilan tensor x yang sudah ada menjadi tensor dengan dimensi baru.

x.reshape(1, 7) adalah operasi yang mengubah bentuk (shape) dari tensor x dari bentuk awalnya (sebuah tensor satu dimensi) menjadi sebuah tensor dua dimensi. Dalam operasi ini, 1 mengindikasikan jumlah baris (atau dimensi pertama) yang diinginkan, dan 7 mengindikasikan jumlah kolom (atau dimensi kedua) yang diinginkan untuk tensor baru.

x_reshaped adalah hasil dari operasi tersebut, dan akan berisi nilai-nilai dari tensor x yang disusun dalam format dua dimensi. Dalam konteks ini, x_reshaped akan menjadi tensor dengan bentuk (1, 7), yang mengindikasikan bahwa tensor tersebut memiliki satu baris dan tujuh kolom.

x_reshaped.shape digunakan untuk memeriksa bentuk (shape) dari tensor x_reshaped. Hasilnya adalah (1, 7), yang menunjukkan bahwa x_reshaped adalah tensor dua dimensi dengan satu baris dan tujuh kolom.

Dengan menggunakan metode .reshape(), kode tersebut menghasilkan tensor x_reshaped yang merupakan representasi ulang dari tensor x ke dalam bentuk dua dimensi dengan satu baris dan tujuh kolom.

We can also change the view with `torch.view()`.

In [ ]:
# Change view (keeps same data as original but changes view)
# See more: https://stackoverflow.com/a/54507446/7900723
z = x.view(1, 7)
z, z.shape

(tensor([[1., 2., 3., 4., 5., 6., 7.]]), torch.Size([1, 7]))

Potongan kode ini menggunakan metode .view() dari PyTorch untuk mengubah tampilan (view) dari tensor x menjadi tensor baru dengan bentuk yang berbeda, tetapi tetap memiliki data yang sama.

x.view(1, 7) adalah operasi yang mengubah bentuk (shape) dari tensor x ke dalam bentuk yang baru. Angka 1 menunjukkan jumlah baris (atau dimensi pertama) yang diinginkan, dan 7 menunjukkan jumlah kolom (atau dimensi kedua) yang diinginkan untuk tensor yang baru.

z adalah hasil dari operasi tersebut, yang akan berisi nilai-nilai yang sama seperti tensor x, tetapi dalam tampilan yang diubah menjadi tensor dengan bentuk (1, 7).

z.shape digunakan untuk memeriksa bentuk (shape) dari tensor z. Hasilnya adalah (1, 7), menunjukkan bahwa z adalah tensor dua dimensi dengan satu baris dan tujuh kolom, sama seperti x_reshaped pada contoh sebelumnya.

Perlu dicatat bahwa metode .view() dalam PyTorch memungkinkan untuk mengubah tampilan tensor tanpa mengubah datanya. Ini berguna saat ingin mengubah bentuk tensor tanpa mengubah isi atau nilai-nilai yang ada dalam tensor. Dalam konteks ini, z adalah representasi ulang dari x ke dalam bentuk tensor dua dimensi dengan satu baris dan tujuh kolom, tetapi data atau nilai-nilai di dalamnya tetap sama dengan tensor x.

In [ ]:
# Changing z changes x
z[:, 0] = 5
z, x

(tensor([[5., 2., 3., 4., 5., 6., 7.]]), tensor([5., 2., 3., 4., 5., 6., 7.]))

In [ ]:
# Stack tensors on top of each other
x_stacked = torch.stack([x, x, x, x], dim=0) # try changing dim to dim=1 and see what happens
x_stacked

tensor([[5., 2., 3., 4., 5., 6., 7.],
        [5., 2., 3., 4., 5., 6., 7.],
        [5., 2., 3., 4., 5., 6., 7.],
        [5., 2., 3., 4., 5., 6., 7.]])

Potongan kode ini mencoba untuk mengilustrasikan hubungan antara tensor z yang merupakan hasil dari operasi .view() terhadap tensor x, serta dampak dari perubahan nilai pada tensor z terhadap tensor x yang asli.

z[:, 0] = 5 adalah operasi yang mengubah nilai pada kolom pertama (indeks 0) dari tensor z menjadi nilai 5 untuk setiap elemen di kolom tersebut.

Setelah operasi tersebut dieksekusi, z akan menjadi tensor yang nilainya diubah menjadi 5 pada kolom pertama, tetapi yang penting untuk dicatat adalah bahwa ini juga mempengaruhi tensor x. Hal ini terjadi karena tensor z sebenarnya adalah tampilan (view) dari tensor x. Ketika menggunakan .view() untuk membuat z, keduanya masih memiliki referensi ke memori yang sama, sehingga perubahan yang dilakukan pada z juga tercermin pada x.

Hasil cetakan z menunjukkan perubahan nilai menjadi tensor([[5., 2., 3., 4., 5., 6., 7.]]). Dan ketika x dicetak, nilai-nilai dalam tensor tersebut juga akan terpengaruh, menampilkan hasil yang sama dengan z.

Jadi, dalam konteks ini, mengubah nilai di tensor z yang merupakan tampilan dari tensor x juga mempengaruhi nilai-nilai di tensor x asli karena keduanya berbagi memori yang sama. Hal ini mengilustrasikan konsep bahwa perubahan yang terjadi pada tensor yang merupakan view dari tensor lainnya juga akan tercermin pada tensor aslinya karena keduanya saling terhubung dalam memori yang sama.

In [ ]:
print(f"Previous tensor: {x_reshaped}")
print(f"Previous shape: {x_reshaped.shape}")

# Remove extra dimension from x_reshaped
x_squeezed = x_reshaped.squeeze()
print(f"\nNew tensor: {x_squeezed}")
print(f"New shape: {x_squeezed.shape}")

Previous tensor: tensor([[5., 2., 3., 4., 5., 6., 7.]])
Previous shape: torch.Size([1, 7])

New tensor: tensor([5., 2., 3., 4., 5., 6., 7.])
New shape: torch.Size([7])


Potongan kode ini memperlihatkan penggunaan metode .squeeze() pada tensor x_reshaped untuk mengurangi dimensi yang tidak perlu, seperti dimensi yang hanya memiliki panjang 1.

x_reshaped adalah tensor hasil dari operasi sebelumnya yang diubah menjadi bentuk dua dimensi (1, 7) menggunakan metode .reshape() atau .view().

x_reshaped.shape mengembalikan bentuk tensor x_reshaped sebelum dilakukan perubahan, yang merupakan tuple (1, 7) menunjukkan bahwa tensor tersebut memiliki satu baris dan tujuh kolom.

x_squeezed = x_reshaped.squeeze() adalah operasi yang menghasilkan tensor baru x_squeezed dengan menghilangkan dimensi yang memiliki panjang 1 dari tensor x_reshaped. Dalam kasus ini, dimensi pertama yang memiliki panjang 1 dihilangkan sehingga tensor yang dihasilkan memiliki bentuk (7,), yang menunjukkan bahwa tensor tersebut merupakan tensor satu dimensi dengan panjang 7.

Setelah operasi .squeeze() dieksekusi, hasil cetakan x_squeezed menunjukkan tensor yang memiliki bentuk (7,), menunjukkan bahwa tensor tersebut telah kehilangan dimensi yang hanya memiliki panjang 1. Sedangkan hasil cetakan x_reshaped tetap menampilkan bentuk (1, 7) dari tensor sebelum di-squeeze.

Penggunaan .squeeze() berguna ketika ingin menghilangkan dimensi yang tidak diperlukan, seperti ketika bekerja dengan tensor yang memiliki dimensi yang tidak relevan atau memiliki panjang 1. Hal ini mempermudah operasi tensor dan mengurangi kompleksitas dimensi yang tidak diperlukan dalam komputasi. Dalam hal ini, x_squeezed adalah hasil dari x_reshaped yang telah di-squeeze, sehingga menghasilkan tensor satu dimensi dengan bentuk (7,) yang lebih sederhana.

In [ ]:
print(f"Previous tensor: {x_squeezed}")
print(f"Previous shape: {x_squeezed.shape}")

## Add an extra dimension with unsqueeze
x_unsqueezed = x_squeezed.unsqueeze(dim=0)
print(f"\nNew tensor: {x_unsqueezed}")
print(f"New shape: {x_unsqueezed.shape}")

Previous tensor: tensor([5., 2., 3., 4., 5., 6., 7.])
Previous shape: torch.Size([7])

New tensor: tensor([[5., 2., 3., 4., 5., 6., 7.]])
New shape: torch.Size([1, 7])


Potongan kode ini menggunakan metode .unsqueeze() pada tensor x_squeezed untuk menambah dimensi baru ke dalam tensor yang telah memiliki bentuk yang lebih sederhana.

x_squeezed adalah tensor hasil dari operasi sebelumnya yang telah di-squeeze sehingga memiliki bentuk satu dimensi (7,).

x_squeezed.shape memberikan bentuk (shape) dari tensor x_squeezed sebelum dilakukan perubahan, yaitu (7,), menunjukkan bahwa tensor tersebut merupakan tensor satu dimensi dengan panjang 7.

x_unsqueezed = x_squeezed.unsqueeze(dim=0) adalah operasi yang menambahkan dimensi baru ke tensor x_squeezed. Dalam kasus ini, dimensi baru dengan panjang 1 ditambahkan di dimensi pertama (dengan parameter dim=0). Hasilnya adalah tensor dengan bentuk (1, 7), menunjukkan bahwa tensor tersebut memiliki satu baris dan tujuh kolom.

Setelah operasi .unsqueeze() dieksekusi, hasil cetakan x_unsqueezed menampilkan tensor yang memiliki bentuk (1, 7), menunjukkan bahwa dimensi baru telah ditambahkan ke dalam tensor. Sedangkan hasil cetakan x_squeezed tetap menunjukkan bentuk (7,) dari tensor sebelum dilakukan penambahan dimensi.

Penggunaan .unsqueeze() berguna ketika ingin menambah dimensi baru ke dalam tensor, terutama saat bekerja dengan tensor yang perlu dimodifikasi bentuknya untuk operasi tertentu, seperti operasi pada dimensi spesifik atau operasi yang memerlukan bentuk tensor tertentu. Dalam hal ini, x_unsqueezed adalah hasil dari x_squeezed yang telah ditambah dimensi sehingga menghasilkan tensor dengan bentuk (1, 7).

In [ ]:
# Create tensor with specific shape
x_original = torch.rand(size=(224, 224, 3))

# Permute the original tensor to rearrange the axis order
x_permuted = x_original.permute(2, 0, 1) # shifts axis 0->1, 1->2, 2->0

print(f"Previous shape: {x_original.shape}")
print(f"New shape: {x_permuted.shape}")

Previous shape: torch.Size([224, 224, 3])
New shape: torch.Size([3, 224, 224])



Potongan kode ini menggunakan metode .permute() pada tensor x_original untuk mengubah urutan sumbu atau axis dari tensor tersebut.

x_original = torch.rand(size=(224, 224, 3)) menciptakan tensor dengan bentuk awal (224, 224, 3). Hal ini menunjukkan bahwa tensor tersebut memiliki dimensi 224 pada sumbu pertama dan kedua, serta memiliki dimensi 3 pada sumbu ketiga.

x_original.shape memberikan informasi tentang bentuk (shape) dari tensor x_original sebelum perubahan, yaitu (224, 224, 3). Hal ini menunjukkan bahwa tensor tersebut memiliki dimensi 224 pada sumbu pertama dan kedua, serta dimensi 3 pada sumbu ketiga.

x_permuted = x_original.permute(2, 0, 1) adalah operasi yang menggunakan metode .permute() untuk mengubah urutan sumbu atau axis pada tensor x_original. Dalam kasus ini, urutan sumbu yang baru adalah (2, 0, 1), yang mengubah posisi sumbu ketiga menjadi sumbu pertama, sumbu pertama menjadi sumbu kedua, dan sumbu kedua menjadi sumbu ketiga.

Setelah operasi .permute() dieksekusi, hasil cetakan x_permuted menunjukkan tensor yang telah mengalami perubahan urutan sumbu. Bentuk tensor baru adalah (3, 224, 224), yang menandakan bahwa sumbu pertama memiliki panjang 3, sumbu kedua memiliki panjang 224, dan sumbu ketiga memiliki panjang 224.

Penggunaan .permute() berguna ketika ingin mengubah urutan sumbu atau axis dari tensor. Ini dapat diperlukan saat bekerja dengan data seperti citra, di mana perubahan urutan sumbu dapat berguna untuk mempersiapkan data untuk pemrosesan atau model tertentu yang memerlukan urutan sumbu yang berbeda. Dalam hal ini, x_permuted adalah hasil dari x_original yang telah mengalami perubahan urutan sumbu.

In [ ]:
# Create a tensor
import torch
x = torch.arange(1, 10).reshape(1, 3, 3)
x, x.shape

(tensor([[[1, 2, 3],
          [4, 5, 6],
          [7, 8, 9]]]),
 torch.Size([1, 3, 3]))

Potongan kode ini membuat tensor x menggunakan fungsi torch.arange() untuk menghasilkan nilai dari 1 hingga 9, kemudian diubah bentuknya menjadi tensor tiga dimensi dengan ukuran (1, 3, 3).

torch.arange(1, 10) menghasilkan sebuah tensor yang berisi nilai dari 1 hingga 9 (inklusif untuk 1, eksklusif untuk 10), membentuk sebuah tensor satu dimensi.

.reshape(1, 3, 3) adalah operasi yang mengubah bentuk (shape) dari tensor yang dihasilkan sebelumnya menjadi bentuk tiga dimensi (1, 3, 3). Dalam kasus ini, 1 mengindikasikan jumlah batch (atau jumlah contoh dalam batch), dan kedua angka 3 mengindikasikan jumlah baris dan kolom yang diinginkan untuk setiap batch.

Hasil cetakan x menunjukkan tensor yang terbentuk, yaitu tensor tiga dimensi dengan bentuk (1, 3, 3). Ini menandakan bahwa tensor tersebut adalah satu batch data dengan masing-masing sampel dalam batch tersebut memiliki tiga baris dan tiga kolom.

Jadi, x adalah tensor tiga dimensi yang mewakili satu batch data dengan ukuran (1, 3, 3). Hal ini bisa digunakan dalam konteks pengolahan data dalam deep learning di mana satu batch data memiliki tiga baris dan tiga kolom untuk setiap sampelnya.

Indexing values goes outer dimension -> inner dimension (check out the square brackets).

In [ ]:
# Let's index bracket by bracket
print(f"First square bracket:\n{x[0]}")
print(f"Second square bracket: {x[0][0]}")
print(f"Third square bracket: {x[0][0][0]}")

First square bracket:
tensor([[1, 2, 3],
        [4, 5, 6],
        [7, 8, 9]])
Second square bracket: tensor([1, 2, 3])
Third square bracket: 1


Potongan kode tersebut memanfaatkan fungsi torch.arange() untuk membuat tensor x dengan nilai dari 1 hingga 9, yang kemudian diubah bentuknya menjadi tensor tiga dimensi dengan ukuran (1, 3, 3).

torch.arange(1, 10) menghasilkan tensor satu dimensi yang berisi nilai dari 1 hingga 9 (inklusif untuk 1, eksklusif untuk 10).

.reshape(1, 3, 3) adalah operasi yang mengubah bentuk tensor hasil dari torch.arange() menjadi tensor tiga dimensi dengan bentuk (1, 3, 3). Angka 1 mengindikasikan jumlah batch (atau contoh dalam batch), sementara angka 3 menunjukkan jumlah baris dan kolom yang diinginkan untuk setiap contoh dalam batch.

Hasil cetakan tensor x menampilkan tensor tiga dimensi dengan bentuk (1, 3, 3), yang menunjukkan bahwa tensor tersebut mewakili satu batch data dengan satu contoh dalam batch. Setiap contoh dalam batch tersebut memiliki tiga baris dan tiga kolom.

Tensor seperti x ini berguna dalam konteks pengolahan data pada deep learning, di mana data seringkali diorganisir dalam batch. Dalam hal ini, tensor x dapat digunakan sebagai batch data dengan satu contoh yang memiliki tiga baris dan tiga kolom.








In [ ]:
# Get all values of 0th dimension and the 0 index of 1st dimension
x[:, 0]

tensor([[1, 2, 3]])

Potongan kode x[:, 0] menggunakan indexing pada tensor x untuk memperoleh subset dari tensor dengan mengambil semua nilai dari dimensi ke-0 dan nilai pada indeks ke-0 dari dimensi ke-1.

Dalam hal ini:

x[:, 0] mengakses tensor x dengan menggunakan : pada dimensi ke-0, yang berarti mengambil semua nilai dari dimensi pertama. Kemudian, 0 pada dimensi ke-1 menunjukkan pengambilan nilai pada indeks ke-0 dari dimensi kedua.
Hasilnya adalah subset dari tensor x yang terdiri dari nilai-nilai pada indeks ke-0 dari dimensi ke-1 (kolom pertama) dari setiap contoh dalam batch (dimensi ke-0). Sebagai contoh, jika x adalah sebuah batch dengan tiga contoh, maka x[:, 0] akan mengambil nilai pertama dari setiap contoh dalam batch tersebut pada dimensi kedua (kolom pertama). Subset ini akan memiliki bentuk (1, 3) jika x adalah sebuah batch dengan satu contoh yang memiliki tiga baris, atau (3,) jika x adalah sebuah batch dengan tiga contoh yang masing-masing memiliki satu baris.

In [ ]:
# Get all values of 0th & 1st dimensions but only index 1 of 2nd dimension
x[:, :, 1]

tensor([[2, 5, 8]])


Potongan kode x[:, :, 1] menggunakan indexing pada tensor x untuk mengambil subset yang mencakup semua nilai dari dimensi ke-0 dan ke-1, tetapi hanya nilai pada indeks ke-1 dari dimensi ke-2.

Dalam konteks ini:

x[:, :, 1] mengakses tensor x dengan menggunakan : pada dimensi ke-0 dan ke-1, yang berarti mengambil semua nilai dari kedua dimensi pertama. Kemudian, 1 pada dimensi ke-2 menunjukkan pengambilan nilai pada indeks ke-1 dari dimensi ketiga.
Hasilnya adalah subset dari tensor x yang terdiri dari nilai-nilai pada indeks ke-1 dari dimensi ketiga (kolom kedua jika diindeks dari nol) dari setiap contoh dalam batch (dimensi ke-0 dan ke-1). Misalnya, jika x adalah sebuah batch dengan tiga contoh, x[:, :, 1] akan mengambil nilai-nilai pada indeks ke-1 dari dimensi ketiga (kolom kedua) dari setiap contoh dalam batch tersebut. Subset ini akan memiliki bentuk (1, 3) jika x adalah sebuah batch dengan satu contoh yang memiliki tiga baris, atau (3,) jika x adalah sebuah batch dengan tiga contoh yang masing-masing memiliki satu baris.

In [ ]:
# Get all values of the 0 dimension but only the 1 index value of the 1st and 2nd dimension
x[:, 1, 1]

tensor([5])


Potongan kode x[:, 1, 1] menggunakan indexing pada tensor x untuk memperoleh subset yang terdiri dari semua nilai dari dimensi ke-0, tetapi hanya nilai pada indeks ke-1 dari dimensi ke-1 dan ke-2.

Dalam konteks ini:

x[:, 1, 1] melakukan akses pada tensor x. : pada dimensi ke-0 berarti mengambil semua nilai dari dimensi pertama. 1 pada dimensi ke-1 menunjukkan pengambilan nilai pada indeks ke-1 dari dimensi kedua. Kemudian, 1 pada dimensi ke-2 menunjukkan pengambilan nilai pada indeks ke-1 dari dimensi ketiga.
Hasilnya adalah subset dari tensor x yang terdiri dari nilai-nilai yang ada pada indeks ke-1 dari dimensi kedua dan ketiga (baris dan kolom kedua jika diindeks dari nol) dari setiap contoh dalam batch pada dimensi ke-0. Misalnya, jika x adalah sebuah batch dengan tiga contoh, x[:, 1, 1] akan mengambil nilai-nilai pada indeks ke-1 dari dimensi kedua dan ketiga dari setiap contoh dalam batch tersebut. Subset ini akan memiliki bentuk (1,) jika x adalah sebuah batch dengan satu contoh, atau (3,) jika x adalah sebuah batch dengan tiga contoh.

In [ ]:
# Get index 0 of 0th and 1st dimension and all values of 2nd dimension
x[0, 0, :] # same as x[0][0]

tensor([1, 2, 3])

Potongan kode x[0, 0, :] atau x[0][0] menggunakan indexing pada tensor x untuk memperoleh subset yang terdiri dari nilai-nilai dari indeks ke-0 dari dimensi ke-0 dan ke-1, dan mengambil semua nilai dari dimensi ke-2.

Dalam konteks ini:

x[0, 0, :] atau x[0][0] mengakses tensor x. 0 pada dimensi ke-0 dan ke-1 menunjukkan pengambilan nilai pada indeks ke-0 dari kedua dimensi tersebut. : pada dimensi ke-2 mengindikasikan pengambilan semua nilai dari dimensi tersebut.
Hasilnya adalah subset dari tensor x yang terdiri dari semua nilai pada dimensi ke-2 (semua nilai dalam baris atau kolom) dari indeks ke-0 dari dimensi ke-0 dan ke-1. Misalnya, jika x adalah sebuah batch dengan tiga contoh, x[0, 0, :] akan mengambil semua nilai dari dimensi ke-2 dari indeks ke-0 dari dimensi ke-0 dan ke-1 dari setiap contoh dalam batch tersebut. Subset ini akan memiliki bentuk (3,) jika x adalah sebuah batch dengan satu contoh, atau (3, 3) jika x adalah sebuah batch dengan tiga contoh.

In [ ]:
# NumPy array to tensor
import torch
import numpy as np
array = np.arange(1.0, 8.0)
tensor = torch.from_numpy(array)
array, tensor

(array([1., 2., 3., 4., 5., 6., 7.]),
 tensor([1., 2., 3., 4., 5., 6., 7.], dtype=torch.float64))

Potongan kode tersebut melakukan konversi dari array NumPy menjadi tensor PyTorch menggunakan fungsi-fungsi dari kedua library tersebut.

np.arange(1.0, 8.0) merupakan fungsi dari NumPy yang menghasilkan array dengan nilai dari 1.0 hingga 7.0 (inklusif untuk 1.0, eksklusif untuk 8.0).
torch.from_numpy(array) digunakan untuk membuat tensor PyTorch dari array NumPy array. Fungsi ini mengambil array NumPy sebagai argumen dan mengonversinya menjadi tensor PyTorch tanpa melakukan penyalinan data, yang berarti tensor yang dihasilkan akan menggunakan memori yang sama dengan array NumPy.
Jadi, array adalah array NumPy yang berisi nilai dari 1.0 hingga 7.0, sedangkan tensor adalah tensor PyTorch yang memiliki nilai yang sama dengan array NumPy dan memungkinkan penggunaan operasi-operasi tensor yang didukung oleh PyTorch. Kode ini bermanfaat ketika kita ingin berpindah antara NumPy dan PyTorch untuk melakukan operasi pada data dengan library yang berbeda.

In [ ]:
# Change the array, keep the tensor
array = array + 1
array, tensor

(array([2., 3., 4., 5., 6., 7., 8.]),
 tensor([1., 2., 3., 4., 5., 6., 7.], dtype=torch.float64))

Potongan kode tersebut mengubah nilai dalam array NumPy yang telah dibuat sebelumnya, tetapi tetap mempertahankan nilai tensor PyTorch yang dibuat dari array tersebut.

array = array + 1 menambahkan 1 ke setiap elemen dalam array NumPy array yang sebelumnya telah dibuat. Hal ini mengubah nilai-nilai dalam array NumPy dengan cara menambahkan 1 pada setiap elemennya.
Meskipun nilai dalam array NumPy telah diubah, nilai tensor PyTorch (tensor) tidak berubah. Ini disebabkan oleh sifat referensial dari fungsi torch.from_numpy(), yang tidak membuat salinan data dari array NumPy, melainkan menggunakan memori yang sama dengan array NumPy.

Jadi, meskipun array telah dimodifikasi dengan menambahkan 1 ke setiap elemennya, nilai tensor tensor tidak berubah dan tetap memiliki nilai yang sama seperti sebelum perubahan pada array NumPy. Ini menunjukkan bahwa tensor PyTorch yang dihasilkan dari array NumPy menggunakan memori yang sama dengan array NumPy dan tidak terpengaruh oleh perubahan yang terjadi pada array NumPy tersebut.

And if you want to go from PyTorch tensor to NumPy array, you can call `tensor.numpy()`.

In [ ]:
# Tensor to NumPy array
tensor = torch.ones(7) # create a tensor of ones with dtype=float32
numpy_tensor = tensor.numpy() # will be dtype=float32 unless changed
tensor, numpy_tensor

(tensor([1., 1., 1., 1., 1., 1., 1.]),
 array([1., 1., 1., 1., 1., 1., 1.], dtype=float32))

Potongan kode tersebut melakukan konversi dari tensor PyTorch menjadi array NumPy menggunakan fungsi-fungsi dari kedua library tersebut.

torch.ones(7) digunakan untuk membuat tensor PyTorch yang terdiri dari tujuh elemen, di mana setiap elemennya memiliki nilai 1.0. Default datatype untuk tensor ini adalah float32.
tensor.numpy() mengonversi tensor PyTorch tensor menjadi array NumPy numpy_tensor. Fungsi ini melakukan konversi tensor PyTorch menjadi array NumPy tanpa melakukan salinan data, artinya array NumPy yang dihasilkan akan menggunakan memori yang sama dengan tensor PyTorch.
Jadi, tensor adalah tensor PyTorch yang terdiri dari tujuh elemen bernilai 1.0 dengan tipe data float32, sedangkan numpy_tensor adalah array NumPy yang memiliki nilai dan tipe data yang sama dengan tensor PyTorch. Perubahan yang dilakukan pada numpy_tensor akan tercermin pada tensor karena keduanya berbagi memori yang sama. Namun, perhatikan bahwa jika terjadi perubahan tipe data pada numpy_tensor, hal itu tidak secara langsung mempengaruhi tipe data pada tensor PyTorch karena konversi tersebut dapat menghasilkan kopi data

In [ ]:
# Change the tensor, keep the array the same
tensor = tensor + 1
tensor, numpy_tensor

(tensor([2., 2., 2., 2., 2., 2., 2.]),
 array([1., 1., 1., 1., 1., 1., 1.], dtype=float32))

Potongan kode ini menambahkan nilai 1 ke setiap elemen dalam tensor PyTorch yang telah dibuat sebelumnya, sementara tetap mempertahankan nilai dan struktur array NumPy yang telah dihasilkan dari tensor tersebut.

tensor = tensor + 1 menambahkan 1 ke setiap elemen dalam tensor PyTorch tensor. Operasi ini menghasilkan tensor baru dengan nilai-nilai yang sudah dimodifikasi.
Meskipun nilai dalam tensor PyTorch tensor telah diubah, nilai dalam array NumPy numpy_tensor tetap tidak berubah. Ini disebabkan oleh sifat konversi tensor ke array tanpa salinan data. Ketika kita melakukan operasi pada tensor PyTorch (tensor), itu tidak mempengaruhi array NumPy (numpy_tensor) karena keduanya memiliki memori yang terpisah dan tidak saling berbagi referensi langsung.

Jadi, setelah operasi penambahan 1 pada tensor, nilai tensor akan berubah sedangkan nilai numpy_tensor akan tetap sama seperti sebelum perubahan, yaitu nilai yang dihasilkan saat konversi dari tensor sebelumnya.

In [ ]:
import torch

# Create two random tensors
random_tensor_A = torch.rand(3, 4)
random_tensor_B = torch.rand(3, 4)

print(f"Tensor A:\n{random_tensor_A}\n")
print(f"Tensor B:\n{random_tensor_B}\n")
print(f"Does Tensor A equal Tensor B? (anywhere)")
random_tensor_A == random_tensor_B

Tensor A:
tensor([[0.8016, 0.3649, 0.6286, 0.9663],
        [0.7687, 0.4566, 0.5745, 0.9200],
        [0.3230, 0.8613, 0.0919, 0.3102]])

Tensor B:
tensor([[0.9536, 0.6002, 0.0351, 0.6826],
        [0.3743, 0.5220, 0.1336, 0.9666],
        [0.9754, 0.8474, 0.8988, 0.1105]])

Does Tensor A equal Tensor B? (anywhere)


tensor([[False, False, False, False],
        [False, False, False, False],
        [False, False, False, False]])


Potongan kode di atas membuat dua tensor acak menggunakan fungsi torch.rand(). Kedua tensor random_tensor_A dan random_tensor_B memiliki dimensi yang sama yaitu (3, 4), yang berarti masing-masing tensor memiliki tiga baris dan empat kolom.

Kemudian, kode mencoba untuk membandingkan elemen-elemen yang sesuai dari kedua tensor menggunakan operator perbandingan == dengan pernyataan random_tensor_A == random_tensor_B.

Namun, dalam kasus ini, hasilnya adalah sebuah tensor Boolean yang menunjukkan apakah setiap pasangan elemen yang sesuai dari kedua tensor tersebut sama atau tidak. Hasil dari perbandingan ini akan menghasilkan tensor yang berisi nilai True atau False di setiap posisi yang sesuai. Jika nilai elemen pada posisi yang sesuai sama antara kedua tensor, maka nilai di posisi tersebut akan menjadi True, jika tidak, nilainya akan menjadi False.

Penting untuk dicatat bahwa perbandingan ini dilakukan element-wise, artinya perbandingan dilakukan pada setiap pasangan elemen yang memiliki posisi yang sama dalam kedua tensor, bukan secara keseluruhan antara dua tensor. Oleh karena itu, hasilnya adalah tensor boolean yang menunjukkan kecocokan elemen-per-elemen antara kedua tensor tersebut.







In [ ]:
import torch
import random

# # Set the random seed
RANDOM_SEED=42 # try changing this to different values and see what happens to the numbers below
torch.manual_seed(seed=RANDOM_SEED)
random_tensor_C = torch.rand(3, 4)

# Have to reset the seed every time a new rand() is called
# Without this, tensor_D would be different to tensor_C
torch.random.manual_seed(seed=RANDOM_SEED) # try commenting this line out and seeing what happens
random_tensor_D = torch.rand(3, 4)

print(f"Tensor C:\n{random_tensor_C}\n")
print(f"Tensor D:\n{random_tensor_D}\n")
print(f"Does Tensor C equal Tensor D? (anywhere)")
random_tensor_C == random_tensor_D

Tensor C:
tensor([[0.8823, 0.9150, 0.3829, 0.9593],
        [0.3904, 0.6009, 0.2566, 0.7936],
        [0.9408, 0.1332, 0.9346, 0.5936]])

Tensor D:
tensor([[0.8823, 0.9150, 0.3829, 0.9593],
        [0.3904, 0.6009, 0.2566, 0.7936],
        [0.9408, 0.1332, 0.9346, 0.5936]])

Does Tensor C equal Tensor D? (anywhere)


tensor([[True, True, True, True],
        [True, True, True, True],
        [True, True, True, True]])

Potongan kode di atas mencoba membuat dua tensor acak, random_tensor_C dan random_tensor_D, dengan menggunakan fungsi torch.rand(). Sebelum membuat random_tensor_C, sebuah seed acak diberikan ke PyTorch dengan menggunakan torch.manual_seed(seed=RANDOM_SEED). Hal ini bertujuan untuk memastikan bahwa urutan bilangan acak yang dihasilkan oleh PyTorch akan sama setiap kali kode tersebut dieksekusi dengan seed yang sama.

Namun, perhatikan bahwa pada saat membuat random_tensor_D, seed acak untuk PyTorch dikembalikan ke RANDOM_SEED lagi dengan torch.random.manual_seed(seed=RANDOM_SEED) sebelum menciptakan tensor. Ini dilakukan untuk memastikan bahwa kedua tensor yang dihasilkan (random_tensor_C dan random_tensor_D) memiliki urutan bilangan acak yang sama.

Setelah kedua tensor dibuat, kode mencoba untuk membandingkan elemen-elemen yang sesuai dari kedua tensor menggunakan operator perbandingan == dengan pernyataan random_tensor_C == random_tensor_D.

Namun, hasilnya tetap sama seperti sebelumnya. Ini karena, meskipun seed acak untuk PyTorch telah diatur ulang dengan nilai yang sama sebelum membuat random_tensor_D, PyTorch menghasilkan bilangan acak secara independen saat fungsi torch.rand() dipanggil, menghasilkan dua tensor dengan urutan bilangan acak yang berbeda, bahkan ketika seed yang sama digunakan.

Sebagai hasilnya, walaupun seed acak diberikan sebelum pembuatan kedua tensor, hasil perbandingan antara kedua tensor adalah sebuah tensor boolean yang menunjukkan kecocokan elemen-per-elemen antara kedua tensor tersebut, yang pada umumnya akan menghasilkan nilai False di sebagian besar posisi dalam tensor hasil perbandingan.

In [ ]:
!nvidia-smi

Thu Jan  4 09:19:19 2024       
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 535.104.05             Driver Version: 535.104.05   CUDA Version: 12.2     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  Tesla T4                       Off | 00000000:00:04.0 Off |                    0 |
| N/A   42C    P8               9W /  70W |      0MiB / 15360MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+--

Kode !nvidia-smi merupakan sebuah perintah "magic command" yang digunakan dalam lingkungan Jupyter Notebook atau Google Colab untuk menampilkan informasi mengenai GPU yang terpasang pada sistem saat ini.

nvidia-smi sendiri adalah utilitas yang digunakan untuk memantau dan mengelola perangkat GPU NVIDIA yang terpasang pada sistem. Saat perintah ini dijalankan dalam lingkungan seperti Jupyter Notebook atau Google Colab, ini akan menampilkan informasi rinci tentang GPU yang tersedia, termasuk informasi tentang nama GPU, penggunaan memori, suhu, utilitas GPU, dan informasi lainnya.

Jika perintah !nvidia-smi dijalankan di lingkungan yang mendukung GPU, outputnya akan menampilkan tabel yang berisi informasi tentang GPU yang terpasang pada sistem tersebut. Namun, jika sistem tersebut tidak memiliki GPU NVIDIA atau tidak terhubung dengan GPU, perintah tersebut mungkin akan menghasilkan pesan kesalahan yang menyatakan bahwa tidak dapat terhubung dengan driver NVIDIA atau tidak ada GPU yang terdeteksi.

If you don't have a Nvidia GPU accessible, the above will output something like:

```
NVIDIA-SMI has failed because it couldn't communicate with the NVIDIA driver. Make sure that the latest NVIDIA driver is installed and running.
```

In that case, go back up and follow the install steps.

If you do have a GPU, the line above will output something like:

```
Wed Jan 19 22:09:08 2022       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 495.46       Driver Version: 460.32.03    CUDA Version: 11.2     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Tesla P100-PCIE...  Off  | 00000000:00:04.0 Off |                    0 |
| N/A   35C    P0    27W / 250W |      0MiB / 16280MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|  No running processes found                                                 |
+-----------------------------------------------------------------------------+
```

In [ ]:
# Check for GPU
import torch
torch.cuda.is_available()

True

Kode tersebut menggunakan fungsi torch.cuda.is_available() untuk memeriksa ketersediaan penggunaan GPU pada sistem saat ini menggunakan PyTorch.

Fungsi torch.cuda.is_available() mengembalikan nilai boolean (True atau False) yang menunjukkan apakah perangkat keras GPU dapat digunakan atau tidak dalam lingkungan saat ini. Jika nilai yang dikembalikan adalah True, itu berarti PyTorch dapat menggunakan GPU untuk melakukan komputasi. Sebaliknya, jika nilai yang dikembalikan adalah False, hal itu menunjukkan bahwa GPU tidak tersedia atau tidak dapat diakses dalam lingkungan saat ini.

Pemeriksaan ketersediaan GPU ini berguna saat ingin memanfaatkan GPU untuk mempercepat proses komputasi, terutama dalam konteks pengolahan data besar atau model deep learning yang kompleks. Jika GPU tersedia, PyTorch dapat menggunakan GPU tersebut untuk menjalankan operasi komputasi pada tensor dengan memanfaatkan kecepatan dan kemampuan paralelisme yang dimiliki oleh perangkat keras GPU.

In [ ]:
# Set device type
device = "cuda" if torch.cuda.is_available() else "cpu"
device

'cuda'

Kode ini menggunakan struktur kondisional untuk menentukan jenis perangkat yang akan digunakan dalam proses komputasi dengan menggunakan PyTorch. Variabel device diinisialisasi dengan nilai "cuda" jika fungsi torch.cuda.is_available() mengembalikan True, yang menunjukkan ketersediaan perangkat keras GPU (CUDA) untuk digunakan oleh PyTorch. Jika tidak, variabel device diatur menjadi "cpu", yang menunjukkan bahwa proses komputasi akan menggunakan unit pemrosesan CPU.

Dalam konteks deep learning dan pengolahan tensor, penggunaan GPU (CUDA) dapat meningkatkan kecepatan komputasi secara signifikan karena kemampuan paralelisme yang dimilikinya. Namun, jika GPU tidak tersedia atau tidak dapat diakses, penggunaan CPU masih memungkinkan untuk menjalankan operasi PyTorch, meskipun dengan kinerja yang lebih lambat dibandingkan dengan penggunaan GPU. Dengan pendekatan ini, kode tersebut memberikan fleksibilitas dalam menentukan perangkat keras yang akan digunakan untuk komputasi PyTorch, bergantung pada ketersediaan perangkat keras yang tersedia dalam sistem.

In [ ]:
# Count number of devices
torch.cuda.device_count()

1


Kode torch.cuda.device_count() digunakan untuk menghitung jumlah perangkat keras GPU yang tersedia dan dapat diakses pada sistem saat ini menggunakan PyTorch. Fungsi ini mengembalikan nilai yang menunjukkan jumlah GPU yang terpasang pada sistem.

Pada sistem yang memiliki beberapa GPU, penggunaan torch.cuda.device_count() memungkinkan pengguna untuk mengetahui jumlah total perangkat keras GPU yang tersedia. Informasi ini berguna dalam pengaturan atau penggunaan PyTorch yang memerlukan penyesuaian pada tingkat perangkat keras yang tersedia. Dengan mengetahui jumlah perangkat keras yang ada, pengguna dapat memilih untuk mendistribusikan atau mengelola tugas komputasi pada beberapa perangkat keras GPU untuk mempercepat proses komputasi, terutama dalam konteks pelatihan model deep learning pada dataset yang besar.

In [ ]:
# Create tensor (default on CPU)
tensor = torch.tensor([1, 2, 3])

# Tensor not on GPU
print(tensor, tensor.device)

# Move tensor to GPU (if available)
tensor_on_gpu = tensor.to(device)
tensor_on_gpu

tensor([1, 2, 3]) cpu


tensor([1, 2, 3], device='cuda:0')

Pada potongan kode ini, sebuah tensor tensor dibuat dengan nilai [1, 2, 3] dan secara default ditempatkan pada CPU. Kemudian, fungsi print digunakan untuk menampilkan nilai tensor tersebut beserta informasi tentang perangkat di mana tensor tersebut berada, yang diperoleh melalui atribut device.

Kemudian, dengan menggunakan variabel device yang telah ditentukan sebelumnya sebagai kondisi (device = "cuda" if torch.cuda.is_available() else "cpu"), kode ini menggunakan metode .to() pada tensor tensor untuk memindahkan tensor tersebut ke perangkat yang sesuai (dalam hal ini, GPU jika tersedia atau CPU jika tidak). Hasilnya adalah tensor yang sama (tensor_on_gpu), tetapi sekarang telah dipindahkan ke perangkat yang telah ditentukan, seperti yang ditunjukkan oleh nilai yang dikembalikan saat tensor tersebut dicetak bersama dengan informasi perangkatnya.

In [ ]:
# If tensor is on GPU, can't transform it to NumPy (this will error)
tensor_on_gpu.numpy()

TypeError: ignored

Potongan kode di atas menunjukkan percobaan untuk mengonversi tensor (tensor_on_gpu) yang telah dipindahkan ke perangkat GPU menjadi array NumPy menggunakan metode .numpy(). Namun, ini menghasilkan error TypeError yang menyatakan bahwa tensor yang berada di perangkat GPU (cuda:0) tidak dapat langsung diubah menjadi array NumPy.

Hal ini terjadi karena metode .numpy() tidak dapat langsung diterapkan pada tensor yang berada di perangkat GPU. Untuk mengonversi tensor yang berada di GPU menjadi array NumPy, langkah pertama adalah memindahkan tensor tersebut kembali ke memori host (CPU) menggunakan metode .cpu() untuk kemudian baru dapat mengonversinya menjadi array NumPy.

Solusinya adalah dengan menggunakan metode .cpu() pada tensor (tensor_on_gpu) untuk menyalin tensor dari GPU ke memori host (CPU), dan kemudian menggunakan metode .numpy() untuk mengonversi tensor yang ada di memori host menjadi array NumPy.

Potongan kode di atas mencoba mengonversi tensor (tensor_on_gpu) yang telah dipindahkan ke perangkat GPU menjadi array NumPy menggunakan metode .numpy(). Namun, ini menghasilkan error TypeError yang menyatakan bahwa tensor yang berada di perangkat GPU (cuda:0) tidak dapat langsung diubah menjadi array NumPy.

Hal ini terjadi karena metode .numpy() tidak dapat langsung diterapkan pada tensor yang berada di perangkat GPU. Untuk mengonversi tensor yang berada di GPU menjadi array NumPy, langkah pertama adalah memindahkan tensor tersebut kembali ke memori host (CPU) menggunakan metode .cpu() untuk kemudian baru dapat mengonversinya menjadi array NumPy.

Solusinya adalah dengan menggunakan metode .cpu() pada tensor (tensor_on_gpu) untuk menyalin tensor dari GPU ke memori host (CPU), dan kemudian menggunakan metode .numpy() untuk mengonversi tensor yang ada di memori host menjadi array NumPy.

In [ ]:
# Instead, copy the tensor back to cpu
tensor_back_on_cpu = tensor_on_gpu.cpu().numpy()
tensor_back_on_cpu

array([1, 2, 3])


Potongan kode di atas menggunakan metode .cpu() pada tensor (tensor_on_gpu) yang berada di perangkat GPU untuk menyalin kembali tensor tersebut ke memori host (CPU). Setelah tensor dipindahkan kembali ke CPU, kemudian digunakan metode .numpy() untuk mengonversi tensor yang ada di memori host (CPU) menjadi array NumPy (tensor_back_on_cpu).

Langkah ini diperlukan karena operasi .numpy() tidak dapat diterapkan langsung pada tensor yang berada di perangkat GPU. Dengan menggunakan .cpu(), tensor dipindahkan kembali ke memori host (CPU), di mana kemudian metode .numpy() dapat diaplikasikan untuk mengonversi tensor menjadi array NumPy. Ini memungkinkan penggunaan data tensor dari perangkat GPU untuk proses pengolahan atau analisis lanjutan yang memerlukan representasi dalam bentuk array NumPy.

In [ ]:
tensor_on_gpu

tensor([1, 2, 3], device='cuda:0')

tensor_on_gpu adalah variabel yang menampung tensor yang telah dipindahkan ke perangkat GPU (Graphics Processing Unit). Pada kode tersebut, terdapat proses penggunaan metode .to() untuk memindahkan tensor dari memori CPU ke memori GPU jika GPU tersedia, atau tetap di memori CPU jika GPU tidak tersedia.

Tensor yang dipindahkan ke perangkat GPU biasanya dimanfaatkan dalam komputasi yang memerlukan kecepatan pemrosesan yang lebih tinggi karena GPU memiliki arsitektur paralel yang kuat untuk tugas-tugas seperti pelatihan model deep learning dan operasi matematika yang intensif.

Penamaan variabel tensor_on_gpu mengindikasikan bahwa tensor tersebut berada di memori GPU, dan hal ini sangat berguna untuk mengetahui lokasi tensor saat bekerja dengan PyTorch, karena operasi-operasi pada tensor dapat bervariasi tergantung pada perangkat keras tempat tensor tersebut berada. Jadi, variabel tensor_on_gpu adalah tensor yang telah berhasil dipindahkan dan saat ini berada di memori GPU.